<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.5.57">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">


<title>Caso práctico (MOV10) – Tumbleweed</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { display: inline-block; text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
</style>


<script src="../../site_libs/quarto-nav/quarto-nav.js"></script>
<script src="../../site_libs/clipboard/clipboard.min.js"></script>
<script src="../../site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="../../site_libs/quarto-search/fuse.min.js"></script>
<script src="../../site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="../../">
<script src="../../site_libs/quarto-html/quarto.js"></script>
<script src="../../site_libs/quarto-html/popper.min.js"></script>
<script src="../../site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="../../site_libs/quarto-html/anchor.min.js"></script>
<link href="../../site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="../../site_libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="../../site_libs/bootstrap/bootstrap.min.js"></script>
<link href="../../site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="../../site_libs/bootstrap/bootstrap.min.css" rel="stylesheet" id="quarto-bootstrap" data-mode="light">
<script id="quarto-search-options" type="application/json">{
  "location": "navbar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "end",
  "type": "overlay",
  "limit": 50,
  "keyboard-shortcut": [
    "f",
    "/",
    "s"
  ],
  "show-item-context": false,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-text-placeholder": "",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit",
    "search-label": "Search"
  }
}</script>


<link rel="stylesheet" href="../../styles.css">
</head>

<body class="nav-fixed">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top">
    <nav class="navbar navbar-expand-lg " data-bs-theme="dark">
      <div class="navbar-container container-fluid">
      <div class="navbar-brand-container mx-auto">
    <a class="navbar-brand" href="../../index.html">
    <span class="navbar-title">Tumbleweed</span>
    </a>
  </div>
            <div id="quarto-search" class="" title="Search"></div>
          <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbarCollapse" aria-controls="navbarCollapse" role="menu" aria-expanded="false" aria-label="Toggle navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
  <span class="navbar-toggler-icon"></span>
</button>
          <div class="collapse navbar-collapse" id="navbarCollapse">
            <ul class="navbar-nav navbar-nav-scroll me-auto">
  <li class="nav-item dropdown ">
    <a class="nav-link dropdown-toggle" href="#" id="nav-menu-r" role="link" data-bs-toggle="dropdown" aria-expanded="false">
 <span class="menu-text">R</span>
    </a>
    <ul class="dropdown-menu" aria-labelledby="nav-menu-r">    
        <li>
    <a class="dropdown-item" href="../../R/introduccion.html">
 <span class="dropdown-text">Introducción a R</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../../R/vectores.html">
 <span class="dropdown-text">Vectores</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../../R/listas.html">
 <span class="dropdown-text">Listas</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../../R/dataframes.html">
 <span class="dropdown-text">Data Frames</span></a>
  </li>  
    </ul>
  </li>
  <li class="nav-item dropdown ">
    <a class="nav-link dropdown-toggle" href="#" id="nav-menu-rna-seq" role="link" data-bs-toggle="dropdown" aria-expanded="false">
 <span class="menu-text">RNA-seq</span>
    </a>
    <ul class="dropdown-menu" aria-labelledby="nav-menu-rna-seq">    
        <li>
    <a class="dropdown-item" href="../../rnaseq/deseq/introduccion.html">
 <span class="dropdown-text">Análisis de expresión diferencial</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../../rnaseq/deseq/ejemplo.html">
 <span class="dropdown-text">Caso práctico (MOV10)</span></a>
  </li>  
    </ul>
  </li>
  <li class="nav-item dropdown ">
    <a class="nav-link dropdown-toggle" href="#" id="nav-menu-enriquecimiento-funcional" role="link" data-bs-toggle="dropdown" aria-expanded="false">
 <span class="menu-text">Enriquecimiento funcional</span>
    </a>
    <ul class="dropdown-menu" aria-labelledby="nav-menu-enriquecimiento-funcional">    
        <li>
    <a class="dropdown-item" href="../../enrfun/introduccion.html">
 <span class="dropdown-text">Enriquecimiento Funcional</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../../enrfun/ora.html">
 <span class="dropdown-text">ORA (MOV10)</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../../enrfun/gprofiler.html">
 <span class="dropdown-text">gProfiler (MOV10)</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../../enrfun/gsea.html">
 <span class="dropdown-text">GSEA (MOV10)</span></a>
  </li>  
    </ul>
  </li>
  <li class="nav-item dropdown ">
    <a class="nav-link dropdown-toggle" href="#" id="nav-menu-herramientas" role="link" data-bs-toggle="dropdown" aria-expanded="false">
 <span class="menu-text">Herramientas</span>
    </a>
    <ul class="dropdown-menu" aria-labelledby="nav-menu-herramientas">    
        <li>
    <a class="dropdown-item" href="../../tools/ensembl.html">
 <span class="dropdown-text">Convertir Ensembl IDs</span></a>
  </li>  
    </ul>
  </li>
  <li class="nav-item dropdown ">
    <a class="nav-link dropdown-toggle" href="#" id="nav-menu-presentaciones" role="link" data-bs-toggle="dropdown" aria-expanded="false">
 <span class="menu-text">Presentaciones</span>
    </a>
    <ul class="dropdown-menu" aria-labelledby="nav-menu-presentaciones">    
        <li>
    <a class="dropdown-item" href="../../presentaciones/usaR/usaR.html">
 <span class="dropdown-text">Usa R!</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../../presentaciones/ggplot2/ggplot2.html">
 <span class="dropdown-text">ggplot2</span></a>
  </li>  
    </ul>
  </li>
</ul>
          </div> <!-- /navcollapse -->
            <div class="quarto-navbar-tools">
</div>
      </div> <!-- /container-fluid -->
    </nav>
</header>
<!-- content -->
<div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article page-navbar">
<!-- sidebar -->
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        <nav id="TOC" role="doc-toc" class="toc-active">
    <h2 id="toc-title">On this page</h2>
   
  <ul>
  <li><a href="#análisis-exploratorio-de-los-datos." id="toc-análisis-exploratorio-de-los-datos." class="nav-link active" data-scroll-target="#análisis-exploratorio-de-los-datos.">Análisis exploratorio de los datos.</a>
  <ul class="collapse">
  <li><a href="#cargar-datos" id="toc-cargar-datos" class="nav-link" data-scroll-target="#cargar-datos">Cargar datos</a></li>
  <li><a href="#deseq2dataset" id="toc-deseq2dataset" class="nav-link" data-scroll-target="#deseq2dataset">DESeq2DataSet</a></li>
  <li><a href="#prefiltrado-de-los-datos" id="toc-prefiltrado-de-los-datos" class="nav-link" data-scroll-target="#prefiltrado-de-los-datos">Prefiltrado de los datos</a></li>
  <li><a href="#normalización-de-las-cuentas" id="toc-normalización-de-las-cuentas" class="nav-link" data-scroll-target="#normalización-de-las-cuentas">Normalización de las cuentas</a></li>
  <li><a href="#transformacion-cuentas" id="toc-transformacion-cuentas" class="nav-link" data-scroll-target="#transformacion-cuentas">Transformacion cuentas</a></li>
  <li><a href="#evaluación-de-calidad" id="toc-evaluación-de-calidad" class="nav-link" data-scroll-target="#evaluación-de-calidad">Evaluación de calidad</a></li>
  </ul></li>
  <li><a href="#deseq2" id="toc-deseq2" class="nav-link" data-scroll-target="#deseq2">DESeq2</a></li>
  <li><a href="#wald-test" id="toc-wald-test" class="nav-link" data-scroll-target="#wald-test">Wald test</a></li>
  <li><a href="#contracción-de-los-fc-para-visualización-y-clasificación." id="toc-contracción-de-los-fc-para-visualización-y-clasificación." class="nav-link" data-scroll-target="#contracción-de-los-fc-para-visualización-y-clasificación.">Contracción de los FC para visualización y clasificación.</a></li>
  <li><a href="#genes-diferencialmente-expresados" id="toc-genes-diferencialmente-expresados" class="nav-link" data-scroll-target="#genes-diferencialmente-expresados">Genes diferencialmente expresados</a>
  <ul class="collapse">
  <li><a href="#volcano-plot" id="toc-volcano-plot" class="nav-link" data-scroll-target="#volcano-plot">Volcano plot</a></li>
  <li><a href="#heatmap" id="toc-heatmap" class="nav-link" data-scroll-target="#heatmap">Heatmap</a></li>
  </ul></li>
  <li><a href="#guardar-resultados" id="toc-guardar-resultados" class="nav-link" data-scroll-target="#guardar-resultados">Guardar resultados</a></li>
  <li><a href="#referencias" id="toc-referencias" class="nav-link" data-scroll-target="#referencias">Referencias</a></li>
  </ul>
</nav>
    </div>
<!-- main -->
<main class="content" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default">
<div class="quarto-title">
<h1 class="title">Caso práctico (MOV10)</h1>
</div>



<div class="quarto-title-meta">

    
  
    
  </div>
  


</header>


<p>Para este análisis se usarán varios paquetes de R algunos de los cuales han sido instalados desde los repositorios CRAN y otro de los de Bioconductor.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb1"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(DESeq2)</span>
<span id="cb1-2"><a href="#cb1-2" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(ggplot2)</span>
<span id="cb1-3"><a href="#cb1-3" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(ggrepel)</span>
<span id="cb1-4"><a href="#cb1-4" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(pheatmap)</span>
<span id="cb1-5"><a href="#cb1-5" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(RColorBrewer)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<details class="code-fold">
<summary>Código funciones</summary>
<div class="sourceCode cell-code" id="cb2"><pre class="sourceCode r code-overflow-wrap code-with-copy"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a>createVolcanoPlot <span class="ot">&lt;-</span> <span class="cf">function</span>(d, log2fc, fdr, <span class="at">max.overlaps =</span> <span class="dv">15</span>) {</span>
<span id="cb2-2"><a href="#cb2-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-3"><a href="#cb2-3" aria-hidden="true" tabindex="-1"></a>    colors <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">"UP"</span><span class="ot">=</span><span class="st">"#FC4E07"</span>, <span class="st">"none"</span><span class="ot">=</span><span class="st">"#E7B800"</span>, <span class="st">"DOWN"</span><span class="ot">=</span><span class="st">"#00AFBB"</span>)</span>
<span id="cb2-4"><a href="#cb2-4" aria-hidden="true" tabindex="-1"></a>    d<span class="sc">$</span>class <span class="ot">&lt;-</span> <span class="st">'none'</span></span>
<span id="cb2-5"><a href="#cb2-5" aria-hidden="true" tabindex="-1"></a>    d<span class="sc">$</span>class[d<span class="sc">$</span>log2FoldChange <span class="sc">&gt;=</span> log2fc <span class="sc">&amp;</span> d<span class="sc">$</span>padj <span class="sc">&lt;=</span> fdr] <span class="ot">&lt;-</span> <span class="st">'UP'</span></span>
<span id="cb2-6"><a href="#cb2-6" aria-hidden="true" tabindex="-1"></a>    d<span class="sc">$</span>class[d<span class="sc">$</span>log2FoldChange <span class="sc">&lt;=</span> <span class="sc">-</span><span class="dv">1</span> <span class="sc">*</span> log2fc <span class="sc">&amp;</span> d<span class="sc">$</span>padj <span class="sc">&lt;=</span> fdr] <span class="ot">&lt;-</span> <span class="st">'DOWN'</span></span>
<span id="cb2-7"><a href="#cb2-7" aria-hidden="true" tabindex="-1"></a>    d<span class="sc">$</span>gene <span class="ot">&lt;-</span> <span class="fu">rownames</span>(d)</span>
<span id="cb2-8"><a href="#cb2-8" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb2-9"><a href="#cb2-9" aria-hidden="true" tabindex="-1"></a>    up_num<span class="ot">=</span><span class="fu">nrow</span>(d[d<span class="sc">$</span>class <span class="sc">==</span> <span class="st">"UP"</span>,])</span>
<span id="cb2-10"><a href="#cb2-10" aria-hidden="true" tabindex="-1"></a>    down_num<span class="ot">=</span><span class="fu">nrow</span>(d[d<span class="sc">$</span>class <span class="sc">==</span> <span class="st">"DOWN"</span>,])</span>
<span id="cb2-11"><a href="#cb2-11" aria-hidden="true" tabindex="-1"></a>    p <span class="ot">&lt;-</span> <span class="fu">ggplot</span>(d, <span class="fu">aes</span>(<span class="at">x =</span> log2FoldChange, <span class="at">y =</span> <span class="sc">-</span><span class="dv">1</span> <span class="sc">*</span> <span class="fu">log10</span>(padj), </span>
<span id="cb2-12"><a href="#cb2-12" aria-hidden="true" tabindex="-1"></a>                       <span class="at">shape =</span> class, <span class="at">fill =</span> class, <span class="at">color =</span> class, <span class="at">label =</span> gene)) <span class="sc">+</span></span>
<span id="cb2-13"><a href="#cb2-13" aria-hidden="true" tabindex="-1"></a>                <span class="fu">geom_point</span>(<span class="at">size =</span> <span class="dv">3</span>) <span class="sc">+</span></span>
<span id="cb2-14"><a href="#cb2-14" aria-hidden="true" tabindex="-1"></a>                <span class="fu">scale_fill_manual</span>(<span class="at">values =</span> colors) <span class="sc">+</span></span>
<span id="cb2-15"><a href="#cb2-15" aria-hidden="true" tabindex="-1"></a>                <span class="fu">scale_color_manual</span>(<span class="at">values =</span> <span class="fu">c</span>(<span class="st">'navy'</span>, <span class="st">"#E7B800"</span>, <span class="st">'navy'</span>)) <span class="sc">+</span></span>
<span id="cb2-16"><a href="#cb2-16" aria-hidden="true" tabindex="-1"></a>                <span class="fu">scale_shape_manual</span>(<span class="at">values =</span> <span class="fu">c</span>(<span class="dv">21</span>, <span class="dv">20</span>, <span class="dv">21</span>)) <span class="sc">+</span></span>
<span id="cb2-17"><a href="#cb2-17" aria-hidden="true" tabindex="-1"></a>                <span class="fu">geom_hline</span>(<span class="at">yintercept =</span> <span class="sc">-</span><span class="dv">1</span> <span class="sc">*</span> <span class="fu">log10</span>(fdr), <span class="at">linetype=</span><span class="st">"dashed"</span>, </span>
<span id="cb2-18"><a href="#cb2-18" aria-hidden="true" tabindex="-1"></a>                              <span class="at">color =</span> <span class="st">"black"</span>, <span class="at">size=</span><span class="fl">0.5</span>) <span class="sc">+</span></span>
<span id="cb2-19"><a href="#cb2-19" aria-hidden="true" tabindex="-1"></a>                <span class="fu">geom_vline</span>(<span class="at">xintercept=</span><span class="fu">c</span>(<span class="sc">-</span><span class="dv">1</span> <span class="sc">*</span> log2fc,log2fc), <span class="at">linetype=</span><span class="st">"dashed"</span>, </span>
<span id="cb2-20"><a href="#cb2-20" aria-hidden="true" tabindex="-1"></a>                              <span class="at">color =</span> <span class="st">"black"</span>, <span class="at">size=</span><span class="fl">0.5</span>) <span class="sc">+</span></span>
<span id="cb2-21"><a href="#cb2-21" aria-hidden="true" tabindex="-1"></a>                 <span class="fu">geom_label_repel</span>(<span class="at">data =</span> d[d<span class="sc">$</span>class <span class="sc">%in%</span> <span class="fu">c</span>(<span class="st">'UP'</span>, <span class="st">'DOWN'</span>),],</span>
<span id="cb2-22"><a href="#cb2-22" aria-hidden="true" tabindex="-1"></a>                                  <span class="at">size=</span> <span class="dv">4</span>,</span>
<span id="cb2-23"><a href="#cb2-23" aria-hidden="true" tabindex="-1"></a>                                  <span class="at">fontface=</span><span class="st">"bold"</span>,</span>
<span id="cb2-24"><a href="#cb2-24" aria-hidden="true" tabindex="-1"></a>                                  <span class="at">color=</span><span class="st">"firebrick"</span>,</span>
<span id="cb2-25"><a href="#cb2-25" aria-hidden="true" tabindex="-1"></a>                                  <span class="at">fill =</span> <span class="st">'white'</span>,</span>
<span id="cb2-26"><a href="#cb2-26" aria-hidden="true" tabindex="-1"></a>                                  <span class="at">box.padding=</span><span class="fu">unit</span>(<span class="dv">1</span>, <span class="st">"lines"</span>),</span>
<span id="cb2-27"><a href="#cb2-27" aria-hidden="true" tabindex="-1"></a>                                  <span class="at">point.padding=</span><span class="fu">unit</span>(<span class="fl">0.5</span>, <span class="st">"lines"</span>),</span>
<span id="cb2-28"><a href="#cb2-28" aria-hidden="true" tabindex="-1"></a>                                  <span class="at">segment.colour =</span> <span class="st">"firebrick"</span>, </span>
<span id="cb2-29"><a href="#cb2-29" aria-hidden="true" tabindex="-1"></a>                                  <span class="at">segment.size =</span> <span class="fl">0.5</span>, </span>
<span id="cb2-30"><a href="#cb2-30" aria-hidden="true" tabindex="-1"></a>                                  <span class="at">segment.alpha =</span> <span class="fl">0.5</span>, </span>
<span id="cb2-31"><a href="#cb2-31" aria-hidden="true" tabindex="-1"></a>                                  <span class="at">max.overlaps =</span> max.overlaps) <span class="sc">+</span></span>
<span id="cb2-32"><a href="#cb2-32" aria-hidden="true" tabindex="-1"></a>                <span class="fu">theme_bw</span>() <span class="sc">+</span></span>
<span id="cb2-33"><a href="#cb2-33" aria-hidden="true" tabindex="-1"></a>                <span class="fu">theme</span>(<span class="at">legend.position=</span><span class="st">"top"</span>) <span class="sc">+</span></span>
<span id="cb2-34"><a href="#cb2-34" aria-hidden="true" tabindex="-1"></a>                <span class="fu">annotate</span>(<span class="at">geom =</span> <span class="st">'text'</span>, <span class="at">label =</span> <span class="fu">paste0</span>(<span class="st">'UP: '</span>, up_num), </span>
<span id="cb2-35"><a href="#cb2-35" aria-hidden="true" tabindex="-1"></a>                         <span class="at">x =</span> <span class="cn">Inf</span>, <span class="at">y =</span> <span class="cn">Inf</span>, <span class="at">hjust =</span> <span class="fl">1.1</span>, <span class="at">vjust =</span> <span class="fl">1.5</span>) <span class="sc">+</span></span>
<span id="cb2-36"><a href="#cb2-36" aria-hidden="true" tabindex="-1"></a>                <span class="fu">annotate</span>(<span class="at">geom =</span> <span class="st">'text'</span>, <span class="at">label =</span> <span class="fu">paste0</span>(<span class="st">'DOWN: '</span>, down_num), </span>
<span id="cb2-37"><a href="#cb2-37" aria-hidden="true" tabindex="-1"></a>                         <span class="at">x =</span> <span class="sc">-</span><span class="cn">Inf</span>, <span class="at">y =</span> <span class="cn">Inf</span>, <span class="at">hjust =</span> <span class="sc">-</span><span class="fl">0.1</span>, <span class="at">vjust =</span> <span class="fl">1.5</span>) <span class="sc">+</span> </span>
<span id="cb2-38"><a href="#cb2-38" aria-hidden="true" tabindex="-1"></a>                <span class="fu">labs</span>(<span class="at">color =</span> <span class="st">"class"</span>)</span>
<span id="cb2-39"><a href="#cb2-39" aria-hidden="true" tabindex="-1"></a>    p</span>
<span id="cb2-40"><a href="#cb2-40" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb2-41"><a href="#cb2-41" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-42"><a href="#cb2-42" aria-hidden="true" tabindex="-1"></a>crearHeatMap <span class="ot">&lt;-</span><span class="cf">function</span>(datos, metadatos) {</span>
<span id="cb2-43"><a href="#cb2-43" aria-hidden="true" tabindex="-1"></a>    <span class="fu">pheatmap</span>(<span class="fu">as.matrix</span>(datos),</span>
<span id="cb2-44"><a href="#cb2-44" aria-hidden="true" tabindex="-1"></a>             <span class="at">scale =</span> <span class="st">'row'</span>,</span>
<span id="cb2-45"><a href="#cb2-45" aria-hidden="true" tabindex="-1"></a>             <span class="at">clustering_method =</span> <span class="st">'ward.D2'</span>,</span>
<span id="cb2-46"><a href="#cb2-46" aria-hidden="true" tabindex="-1"></a>             <span class="at">cutree_cols =</span>  <span class="dv">3</span>,</span>
<span id="cb2-47"><a href="#cb2-47" aria-hidden="true" tabindex="-1"></a>             <span class="at">cutree_rows =</span> <span class="dv">2</span>,</span>
<span id="cb2-48"><a href="#cb2-48" aria-hidden="true" tabindex="-1"></a>             <span class="at">annotation_col =</span> metadatos,</span>
<span id="cb2-49"><a href="#cb2-49" aria-hidden="true" tabindex="-1"></a>             <span class="at">fontsize =</span> <span class="dv">8</span>,</span>
<span id="cb2-50"><a href="#cb2-50" aria-hidden="true" tabindex="-1"></a>             <span class="at">color =</span> <span class="fu">colorRampPalette</span>(<span class="fu">c</span>(<span class="st">"steelblue"</span>, <span class="st">"white"</span>, <span class="st">"firebrick3"</span>))(<span class="dv">100</span>)</span>
<span id="cb2-51"><a href="#cb2-51" aria-hidden="true" tabindex="-1"></a>            )</span>
<span id="cb2-52"><a href="#cb2-52" aria-hidden="true" tabindex="-1"></a>}</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
<section id="análisis-exploratorio-de-los-datos." class="level2">
<h2 class="anchored" data-anchor-id="análisis-exploratorio-de-los-datos.">Análisis exploratorio de los datos.</h2>
<section id="cargar-datos" class="level3">
<h3 class="anchored" data-anchor-id="cargar-datos">Cargar datos</h3>
<p>Para ilustrar el uso de DESeq usaremos el dataset MOV10. Cargamos los datos de la matriz de cuentas sin normalizar:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb3"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb3-1"><a href="#cb3-1" aria-hidden="true" tabindex="-1"></a>df.cts <span class="ot">&lt;-</span> <span class="fu">read.table</span>(<span class="st">"data/Mov10_full_counts.txt"</span>, <span class="at">header=</span>T, <span class="at">row.names=</span><span class="dv">1</span>) </span>
<span id="cb3-2"><a href="#cb3-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-3"><a href="#cb3-3" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(df.cts)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div class="kable-table">
<table class="caption-top table table-sm table-striped small">
<caption>Matriz de cuentas</caption>
<colgroup>
<col style="width: 12%">
<col style="width: 11%">
<col style="width: 11%">
<col style="width: 11%">
<col style="width: 11%">
<col style="width: 11%">
<col style="width: 11%">
<col style="width: 11%">
<col style="width: 11%">
</colgroup>
<thead>
<tr class="header">
<th style="text-align: left;"></th>
<th style="text-align: right;">Mov10_kd_2</th>
<th style="text-align: right;">Mov10_kd_3</th>
<th style="text-align: right;">Mov10_oe_1</th>
<th style="text-align: right;">Mov10_oe_2</th>
<th style="text-align: right;">Mov10_oe_3</th>
<th style="text-align: right;">Irrel_kd_1</th>
<th style="text-align: right;">Irrel_kd_2</th>
<th style="text-align: right;">Irrel_kd_3</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td style="text-align: left;">1/2-SBSRNA4</td>
<td style="text-align: right;">57</td>
<td style="text-align: right;">41</td>
<td style="text-align: right;">64</td>
<td style="text-align: right;">55</td>
<td style="text-align: right;">38</td>
<td style="text-align: right;">45</td>
<td style="text-align: right;">31</td>
<td style="text-align: right;">39</td>
</tr>
<tr class="even">
<td style="text-align: left;">A1BG</td>
<td style="text-align: right;">71</td>
<td style="text-align: right;">40</td>
<td style="text-align: right;">100</td>
<td style="text-align: right;">81</td>
<td style="text-align: right;">41</td>
<td style="text-align: right;">77</td>
<td style="text-align: right;">58</td>
<td style="text-align: right;">40</td>
</tr>
<tr class="odd">
<td style="text-align: left;">A1BG-AS1</td>
<td style="text-align: right;">256</td>
<td style="text-align: right;">177</td>
<td style="text-align: right;">220</td>
<td style="text-align: right;">189</td>
<td style="text-align: right;">107</td>
<td style="text-align: right;">213</td>
<td style="text-align: right;">172</td>
<td style="text-align: right;">126</td>
</tr>
<tr class="even">
<td style="text-align: left;">A1CF</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
</tr>
<tr class="odd">
<td style="text-align: left;">A2LD1</td>
<td style="text-align: right;">146</td>
<td style="text-align: right;">81</td>
<td style="text-align: right;">138</td>
<td style="text-align: right;">125</td>
<td style="text-align: right;">52</td>
<td style="text-align: right;">91</td>
<td style="text-align: right;">80</td>
<td style="text-align: right;">50</td>
</tr>
<tr class="even">
<td style="text-align: left;">A2M</td>
<td style="text-align: right;">10</td>
<td style="text-align: right;">9</td>
<td style="text-align: right;">2</td>
<td style="text-align: right;">5</td>
<td style="text-align: right;">2</td>
<td style="text-align: right;">9</td>
<td style="text-align: right;">8</td>
<td style="text-align: right;">4</td>
</tr>
</tbody>
</table>
</div>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb4"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb4-1"><a href="#cb4-1" aria-hidden="true" tabindex="-1"></a>metadata <span class="ot">&lt;-</span> <span class="fu">read.table</span>(<span class="st">"data/Mov10_full_meta.txt"</span>, <span class="at">header=</span>T, <span class="at">row.names=</span><span class="dv">1</span>) </span>
<span id="cb4-2"><a href="#cb4-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb4-3"><a href="#cb4-3" aria-hidden="true" tabindex="-1"></a>metadata</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div class="kable-table">
<table class="caption-top table table-sm table-striped small">
<caption>Metadatos</caption>
<thead>
<tr class="header">
<th style="text-align: left;"></th>
<th style="text-align: left;">sampletype</th>
<th style="text-align: left;">MOVexpr</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td style="text-align: left;">Mov10_kd_2</td>
<td style="text-align: left;">MOV10_knockdown</td>
<td style="text-align: left;">low</td>
</tr>
<tr class="even">
<td style="text-align: left;">Mov10_kd_3</td>
<td style="text-align: left;">MOV10_knockdown</td>
<td style="text-align: left;">low</td>
</tr>
<tr class="odd">
<td style="text-align: left;">Mov10_oe_1</td>
<td style="text-align: left;">MOV10_overexpression</td>
<td style="text-align: left;">high</td>
</tr>
<tr class="even">
<td style="text-align: left;">Mov10_oe_2</td>
<td style="text-align: left;">MOV10_overexpression</td>
<td style="text-align: left;">high</td>
</tr>
<tr class="odd">
<td style="text-align: left;">Mov10_oe_3</td>
<td style="text-align: left;">MOV10_overexpression</td>
<td style="text-align: left;">high</td>
</tr>
<tr class="even">
<td style="text-align: left;">Irrel_kd_1</td>
<td style="text-align: left;">control</td>
<td style="text-align: left;">normal</td>
</tr>
<tr class="odd">
<td style="text-align: left;">Irrel_kd_2</td>
<td style="text-align: left;">control</td>
<td style="text-align: left;">normal</td>
</tr>
<tr class="even">
<td style="text-align: left;">Irrel_kd_3</td>
<td style="text-align: left;">control</td>
<td style="text-align: left;">normal</td>
</tr>
</tbody>
</table>
</div>
</div>
</div>
<p>Es importante comprobar que los nombres de las muestras coinciden en los dos ficheros y estan en el orden correcto.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb5"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb5-1"><a href="#cb5-1" aria-hidden="true" tabindex="-1"></a><span class="fu">all</span>(<span class="fu">names</span>(df.cts) <span class="sc">%in%</span> <span class="fu">rownames</span>(metadata))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] TRUE</code></pre>
</div>
<div class="sourceCode cell-code" id="cb7"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb7-1"><a href="#cb7-1" aria-hidden="true" tabindex="-1"></a><span class="fu">all</span>(<span class="fu">names</span>(df.cts) <span class="sc">==</span> <span class="fu">rownames</span>(metadata))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] TRUE</code></pre>
</div>
</div>
</section>
<section id="deseq2dataset" class="level3">
<h3 class="anchored" data-anchor-id="deseq2dataset">DESeq2DataSet</h3>
<p>Lo primero que necesitamos es crear un objeto DESeqDataSet. DESeqDataSet es una clase personalizada del paquete DESeq2 de Bioconductor. Esta clase se utiliza para almacenar los datos y calculos necesarios. Para crear este objeto necesitamos la matriz de cuentas, los metadatos y una “formula” para el diseño. La formula especifica la columna(o columnas) en la tabla de metadatos que se quiere utilizar para el análisis. - <em>Nota:</em> Es esencial que el nombre de las muestras coincidan en los dos ficheros y esten en el mismo orden.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb9"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb9-1"><a href="#cb9-1" aria-hidden="true" tabindex="-1"></a>dds <span class="ot">&lt;-</span> <span class="fu">DESeqDataSetFromMatrix</span>(</span>
<span id="cb9-2"><a href="#cb9-2" aria-hidden="true" tabindex="-1"></a>                <span class="at">countData =</span> df.cts, </span>
<span id="cb9-3"><a href="#cb9-3" aria-hidden="true" tabindex="-1"></a>                <span class="at">colData =</span> metadata, </span>
<span id="cb9-4"><a href="#cb9-4" aria-hidden="true" tabindex="-1"></a>                <span class="at">design =</span> <span class="sc">~</span>sampletype</span>
<span id="cb9-5"><a href="#cb9-5" aria-hidden="true" tabindex="-1"></a>                )</span>
<span id="cb9-6"><a href="#cb9-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb9-7"><a href="#cb9-7" aria-hidden="true" tabindex="-1"></a>dds</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>class: DESeqDataSet 
dim: 23368 8 
metadata(1): version
assays(1): counts
rownames(23368): 1/2-SBSRNA4 A1BG ... ZZZ3 tAKR
rowData names(0):
colnames(8): Mov10_kd_2 Mov10_kd_3 ... Irrel_kd_2 Irrel_kd_3
colData names(2): sampletype MOVexpr</code></pre>
</div>
</div>
</section>
<section id="prefiltrado-de-los-datos" class="level3">
<h3 class="anchored" data-anchor-id="prefiltrado-de-los-datos">Prefiltrado de los datos</h3>
<p>Tipicamente las matrices de RNA-seq contienen muchas filas con ceros, y adicionalmente otras con valores muy bajos de cuentas. Es recomendable eliminar los genes que no aportan (o casi no aportan ) informacion. Aquí realizaremos un filtrado para mantener los genes con un recuento de al menos 10 para una cantidad mínima de muestras. El valor de 10 es razonamble para datos de secuenciación Bulk-RNA. Una recomendación para el número mínimo de muestras es especificar el tamaño de grupo más pequeño.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb11"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb11-1"><a href="#cb11-1" aria-hidden="true" tabindex="-1"></a><span class="fu">nrow</span>(dds)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 23368</code></pre>
</div>
<div class="sourceCode cell-code" id="cb13"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb13-1"><a href="#cb13-1" aria-hidden="true" tabindex="-1"></a>smallestGroupSize <span class="ot">&lt;-</span> <span class="dv">2</span></span>
<span id="cb13-2"><a href="#cb13-2" aria-hidden="true" tabindex="-1"></a>keep <span class="ot">&lt;-</span> <span class="fu">rowSums</span>(<span class="fu">counts</span>(dds) <span class="sc">&gt;=</span> <span class="dv">10</span>) <span class="sc">&gt;=</span> smallestGroupSize</span>
<span id="cb13-3"><a href="#cb13-3" aria-hidden="true" tabindex="-1"></a>dds <span class="ot">&lt;-</span> dds[keep,]</span>
<span id="cb13-4"><a href="#cb13-4" aria-hidden="true" tabindex="-1"></a><span class="fu">nrow</span>(dds)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 15365</code></pre>
</div>
</div>
</section>
<section id="normalización-de-las-cuentas" class="level3">
<h3 class="anchored" data-anchor-id="normalización-de-las-cuentas">Normalización de las cuentas</h3>
<p>En DESeq2 la normalización se calcula usando el metodo de la mediana de las relaciones.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb15"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb15-1"><a href="#cb15-1" aria-hidden="true" tabindex="-1"></a>dds <span class="ot">&lt;-</span> <span class="fu">estimateSizeFactors</span>(dds)</span>
<span id="cb15-2"><a href="#cb15-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-3"><a href="#cb15-3" aria-hidden="true" tabindex="-1"></a>dds</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>class: DESeqDataSet 
dim: 15365 8 
metadata(1): version
assays(1): counts
rownames(15365): 1/2-SBSRNA4 A1BG ... ZZEF1 ZZZ3
rowData names(0):
colnames(8): Mov10_kd_2 Mov10_kd_3 ... Irrel_kd_2 Irrel_kd_3
colData names(3): sampletype MOVexpr sizeFactor</code></pre>
</div>
</div>
</section>
<section id="transformacion-cuentas" class="level3">
<h3 class="anchored" data-anchor-id="transformacion-cuentas">Transformacion cuentas</h3>
<p>Los métodos estadísticos para análisis exploratorio funcionan mejor con datos que tienen similares rangos de varianza a diferentes rangos de medias. En el caso de las cuentas de RNA-seq la varianza crece con la media, con lo cual el PCA estará influenciado por los genes con numeros de cuentas altos. La estrategía para evitarlo es tomar el logaritmo de las cuentas normalizadas sumandole una pseudocuenta (para los casos de 0 cuentas, el log de 0 es indefinido). Sin embargo, ahora los genes con pocas cuentas pueden dominar los resuldatos. DESeq2 usa una <strong>transformación logaritmica regularizada</strong> como solucion. Para los genes con valores altos el resultado es similar a la transfomación logaritmica log2. Para los genes con valores pequeños los valores se reducen hacia la media de los genes a lo largo de todas las muestras.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb17"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb17-1"><a href="#cb17-1" aria-hidden="true" tabindex="-1"></a>rld <span class="ot">&lt;-</span> <span class="fu">rlog</span>(dds, <span class="at">blind =</span> <span class="cn">TRUE</span>)</span>
<span id="cb17-2"><a href="#cb17-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-3"><a href="#cb17-3" aria-hidden="true" tabindex="-1"></a>rld</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>class: DESeqTransform 
dim: 15365 8 
metadata(1): version
assays(1): ''
rownames(15365): 1/2-SBSRNA4 A1BG ... ZZEF1 ZZZ3
rowData names(7): baseMean baseVar ... dispFit rlogIntercept
colnames(8): Mov10_kd_2 Mov10_kd_3 ... Irrel_kd_2 Irrel_kd_3
colData names(3): sampletype MOVexpr sizeFactor</code></pre>
</div>
</div>
</section>
<section id="evaluación-de-calidad" class="level3">
<h3 class="anchored" data-anchor-id="evaluación-de-calidad">Evaluación de calidad</h3>
<section id="análisis-de-componentes-principales-pca" class="level4">
<h4 class="anchored" data-anchor-id="análisis-de-componentes-principales-pca">Análisis de componentes principales (PCA)</h4>
<p>Una manera de visualizar las distancias muestra a muestra es el PCA.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb19"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb19-1"><a href="#cb19-1" aria-hidden="true" tabindex="-1"></a><span class="fu">plotPCA</span>(rld, <span class="at">intgroup=</span><span class="fu">c</span>(<span class="st">'sampletype'</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div id="fig-pca" class="quarto-float quarto-figure quarto-figure-center anchored">
<figure class="quarto-float quarto-float-fig figure">
<div aria-describedby="fig-pca-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<img src="ejemplo_files/figure-html/fig-pca-1.png" class="img-fluid figure-img" width="672">
</div>
<figcaption class="quarto-float-caption-bottom quarto-float-caption quarto-float-fig" id="fig-pca-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Figure&nbsp;1: Análisis de componentes principales
</figcaption>
</figure>
</div>
</div>
</div>
</section>
<section id="clustering-jerarquico" class="level4">
<h4 class="anchored" data-anchor-id="clustering-jerarquico">Clustering jerarquico</h4>
<p>Otra metodo es agrupar las muestras basadas en su similitud. En este caso se puede usar la correlación de Pearson para medir la silitud entre muestras. Generealmente se espera una correlación bastante alta (&gt;0.95) entre todas las muestras, además se esperan ver las muestras agrupadas de manera similar a lo observado en el PCA.</p>
<p>Las muestras que muestran valores bajos de correlación con las demas muestras (&lt;0.80) representan outliers.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb20"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb20-1"><a href="#cb20-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Estraemos la matix rlog del objeto</span></span>
<span id="cb20-2"><a href="#cb20-2" aria-hidden="true" tabindex="-1"></a>rld_mat <span class="ot">&lt;-</span> <span class="fu">assay</span>(rld) </span>
<span id="cb20-3"><a href="#cb20-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb20-4"><a href="#cb20-4" aria-hidden="true" tabindex="-1"></a><span class="co"># Calculamos las correlaciones</span></span>
<span id="cb20-5"><a href="#cb20-5" aria-hidden="true" tabindex="-1"></a>rld_cor <span class="ot">&lt;-</span> <span class="fu">cor</span>(rld_mat)</span>
<span id="cb20-6"><a href="#cb20-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb20-7"><a href="#cb20-7" aria-hidden="true" tabindex="-1"></a><span class="co">#heatmap</span></span>
<span id="cb20-8"><a href="#cb20-8" aria-hidden="true" tabindex="-1"></a><span class="fu">pheatmap</span>(rld_cor, <span class="at">color =</span>  <span class="fu">brewer.pal</span>(<span class="dv">6</span>, <span class="st">"Blues"</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="ejemplo_files/figure-html/pheapmat-correlation-1.png" class="img-fluid figure-img" width="672"></p>
<figcaption>Matriz de correlaciones entre las muestras.</figcaption>
</figure>
</div>
</div>
</div>
</section>
</section>
</section>
<section id="deseq2" class="level2">
<h2 class="anchored" data-anchor-id="deseq2">DESeq2</h2>
<p>Para realizar el calculo ejecutamos la funcion <strong>DESeq()</strong>, que requiere como input un objeto del tipo <strong>DESeqDataSet</strong> (la variable <em>dds</em>). Reasignando el resultado a la misma variable continuamos llenando los <em>slots</em> del objeto DESeqDataSet.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb21"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb21-1"><a href="#cb21-1" aria-hidden="true" tabindex="-1"></a>dds <span class="ot">&lt;-</span> <span class="fu">DESeq</span>(dds)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>using pre-existing size factors</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>estimating dispersions</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>gene-wise dispersion estimates</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>mean-dispersion relationship</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>final dispersion estimates</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>fitting model and testing</code></pre>
</div>
<div class="sourceCode cell-code" id="cb28"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb28-1"><a href="#cb28-1" aria-hidden="true" tabindex="-1"></a>dds</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>class: DESeqDataSet 
dim: 15365 8 
metadata(1): version
assays(4): counts mu H cooks
rownames(15365): 1/2-SBSRNA4 A1BG ... ZZEF1 ZZZ3
rowData names(26): baseMean baseVar ... deviance maxCooks
colnames(8): Mov10_kd_2 Mov10_kd_3 ... Irrel_kd_2 Irrel_kd_3
colData names(3): sampletype MOVexpr sizeFactor</code></pre>
</div>
</div>
</section>
<section id="wald-test" class="level2">
<h2 class="anchored" data-anchor-id="wald-test">Wald test</h2>
<p>La función <strong>results()</strong> devuelve los valores de log2 fold changes y los p-values para el Wald test de la comparación.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb30"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb30-1"><a href="#cb30-1" aria-hidden="true" tabindex="-1"></a>res <span class="ot">&lt;-</span> <span class="fu">results</span>(dds, <span class="at">contrast =</span> <span class="fu">c</span>(<span class="st">'sampletype'</span>, <span class="st">'MOV10_overexpression'</span>, <span class="st">'control'</span>))</span>
<span id="cb30-2"><a href="#cb30-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb30-3"><a href="#cb30-3" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(<span class="fu">as.data.frame</span>(res))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div class="kable-table">
<table class="caption-top table table-sm table-striped small">
<caption>FC y valores p</caption>
<colgroup>
<col style="width: 15%">
<col style="width: 13%">
<col style="width: 18%">
<col style="width: 12%">
<col style="width: 13%">
<col style="width: 12%">
<col style="width: 12%">
</colgroup>
<thead>
<tr class="header">
<th style="text-align: left;"></th>
<th style="text-align: right;">baseMean</th>
<th style="text-align: right;">log2FoldChange</th>
<th style="text-align: right;">lfcSE</th>
<th style="text-align: right;">stat</th>
<th style="text-align: right;">pvalue</th>
<th style="text-align: right;">padj</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td style="text-align: left;">1/2-SBSRNA4</td>
<td style="text-align: right;">45.67985</td>
<td style="text-align: right;">0.3753332</td>
<td style="text-align: right;">0.2597951</td>
<td style="text-align: right;">1.4447276</td>
<td style="text-align: right;">0.1485344</td>
<td style="text-align: right;">0.2442718</td>
</tr>
<tr class="even">
<td style="text-align: left;">A1BG</td>
<td style="text-align: right;">61.13287</td>
<td style="text-align: right;">0.2664358</td>
<td style="text-align: right;">0.2195424</td>
<td style="text-align: right;">1.2135962</td>
<td style="text-align: right;">0.2249020</td>
<td style="text-align: right;">0.3369021</td>
</tr>
<tr class="odd">
<td style="text-align: left;">A1BG-AS1</td>
<td style="text-align: right;">175.74157</td>
<td style="text-align: right;">-0.0562199</td>
<td style="text-align: right;">0.1362360</td>
<td style="text-align: right;">-0.4126656</td>
<td style="text-align: right;">0.6798516</td>
<td style="text-align: right;">0.7715430</td>
</tr>
<tr class="even">
<td style="text-align: left;">A2LD1</td>
<td style="text-align: right;">89.65925</td>
<td style="text-align: right;">0.4218562</td>
<td style="text-align: right;">0.1922733</td>
<td style="text-align: right;">2.1940443</td>
<td style="text-align: right;">0.0282322</td>
<td style="text-align: right;">0.0630414</td>
</tr>
<tr class="odd">
<td style="text-align: left;">A4GALT</td>
<td style="text-align: right;">64.59690</td>
<td style="text-align: right;">1.0013568</td>
<td style="text-align: right;">0.2160868</td>
<td style="text-align: right;">4.6340497</td>
<td style="text-align: right;">0.0000036</td>
<td style="text-align: right;">0.0000241</td>
</tr>
<tr class="even">
<td style="text-align: left;">AAAS</td>
<td style="text-align: right;">1454.28689</td>
<td style="text-align: right;">-0.0430074</td>
<td style="text-align: right;">0.0680939</td>
<td style="text-align: right;">-0.6315892</td>
<td style="text-align: right;">0.5276553</td>
<td style="text-align: right;">0.6432932</td>
</tr>
</tbody>
</table>
</div>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb31"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb31-1"><a href="#cb31-1" aria-hidden="true" tabindex="-1"></a><span class="fu">as.data.frame</span>(<span class="fu">mcols</span>(res, <span class="at">use.names =</span> <span class="cn">TRUE</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div class="kable-table">
<table class="caption-top table table-sm table-striped small">
<caption>Descripción de los resultados</caption>
<colgroup>
<col style="width: 15%">
<col style="width: 13%">
<col style="width: 70%">
</colgroup>
<thead>
<tr class="header">
<th style="text-align: left;"></th>
<th style="text-align: left;">type</th>
<th style="text-align: left;">description</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td style="text-align: left;">baseMean</td>
<td style="text-align: left;">intermediate</td>
<td style="text-align: left;">mean of normalized counts for all samples</td>
</tr>
<tr class="even">
<td style="text-align: left;">log2FoldChange</td>
<td style="text-align: left;">results</td>
<td style="text-align: left;">log2 fold change (MLE): sampletype MOV10_overexpression vs control</td>
</tr>
<tr class="odd">
<td style="text-align: left;">lfcSE</td>
<td style="text-align: left;">results</td>
<td style="text-align: left;">standard error: sampletype MOV10 overexpression vs control</td>
</tr>
<tr class="even">
<td style="text-align: left;">stat</td>
<td style="text-align: left;">results</td>
<td style="text-align: left;">Wald statistic: sampletype MOV10 overexpression vs control</td>
</tr>
<tr class="odd">
<td style="text-align: left;">pvalue</td>
<td style="text-align: left;">results</td>
<td style="text-align: left;">Wald test p-value: sampletype MOV10 overexpression vs control</td>
</tr>
<tr class="even">
<td style="text-align: left;">padj</td>
<td style="text-align: left;">results</td>
<td style="text-align: left;">BH adjusted p-values</td>
</tr>
</tbody>
</table>
</div>
</div>
</div>
<div class="callout callout-style-default callout-note callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Note
</div>
</div>
<div class="callout-body-container callout-body">
<p>Algunos valores de p pueden aparecer como <strong>NA</strong> en la tabla de resultados, esto puede ser debido a que baseMean pueda ser cero, que una fila contenga un valor considerado outlier, o que esa fila sea filtrada por el filtro independiente.</p>
</div>
</div>
</section>
<section id="contracción-de-los-fc-para-visualización-y-clasificación." class="level2">
<h2 class="anchored" data-anchor-id="contracción-de-los-fc-para-visualización-y-clasificación.">Contracción de los FC para visualización y clasificación.</h2>
<p>La reducción del tamaño del efecto (estimaciones de LFC) es útil para la visualización y clasificación de genes</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb32"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb32-1"><a href="#cb32-1" aria-hidden="true" tabindex="-1"></a><span class="fu">resultsNames</span>(dds)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "Intercept"                                 
[2] "sampletype_MOV10_knockdown_vs_control"     
[3] "sampletype_MOV10_overexpression_vs_control"</code></pre>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb34"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb34-1"><a href="#cb34-1" aria-hidden="true" tabindex="-1"></a>resLFC <span class="ot">&lt;-</span> <span class="fu">lfcShrink</span>(dds, <span class="at">coef=</span><span class="st">"sampletype_MOV10_overexpression_vs_control"</span>, <span class="at">type=</span><span class="st">"ashr"</span>)</span>
<span id="cb34-2"><a href="#cb34-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb34-3"><a href="#cb34-3" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(<span class="fu">as.data.frame</span>(resLFC))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div class="kable-table">
<table class="caption-top table table-sm table-striped small">
<caption>FC y valores p</caption>
<colgroup>
<col style="width: 17%">
<col style="width: 16%">
<col style="width: 22%">
<col style="width: 14%">
<col style="width: 14%">
<col style="width: 14%">
</colgroup>
<thead>
<tr class="header">
<th style="text-align: left;"></th>
<th style="text-align: right;">baseMean</th>
<th style="text-align: right;">log2FoldChange</th>
<th style="text-align: right;">lfcSE</th>
<th style="text-align: right;">pvalue</th>
<th style="text-align: right;">padj</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td style="text-align: left;">1/2-SBSRNA4</td>
<td style="text-align: right;">45.67985</td>
<td style="text-align: right;">0.1868058</td>
<td style="text-align: right;">0.1902838</td>
<td style="text-align: right;">0.1485344</td>
<td style="text-align: right;">0.2442718</td>
</tr>
<tr class="even">
<td style="text-align: left;">A1BG</td>
<td style="text-align: right;">61.13287</td>
<td style="text-align: right;">0.1510142</td>
<td style="text-align: right;">0.1684756</td>
<td style="text-align: right;">0.2249020</td>
<td style="text-align: right;">0.3369021</td>
</tr>
<tr class="odd">
<td style="text-align: left;">A1BG-AS1</td>
<td style="text-align: right;">175.74157</td>
<td style="text-align: right;">-0.0423753</td>
<td style="text-align: right;">0.1183626</td>
<td style="text-align: right;">0.6798516</td>
<td style="text-align: right;">0.7715430</td>
</tr>
<tr class="even">
<td style="text-align: left;">A2LD1</td>
<td style="text-align: right;">89.65925</td>
<td style="text-align: right;">0.2728795</td>
<td style="text-align: right;">0.1627217</td>
<td style="text-align: right;">0.0282322</td>
<td style="text-align: right;">0.0630414</td>
</tr>
<tr class="odd">
<td style="text-align: left;">A4GALT</td>
<td style="text-align: right;">64.59690</td>
<td style="text-align: right;">0.7830021</td>
<td style="text-align: right;">0.2372903</td>
<td style="text-align: right;">0.0000036</td>
<td style="text-align: right;">0.0000241</td>
</tr>
<tr class="even">
<td style="text-align: left;">AAAS</td>
<td style="text-align: right;">1454.28689</td>
<td style="text-align: right;">-0.0396237</td>
<td style="text-align: right;">0.0653723</td>
<td style="text-align: right;">0.5276553</td>
<td style="text-align: right;">0.6432932</td>
</tr>
</tbody>
</table>
</div>
</div>
</div>
</section>
<section id="genes-diferencialmente-expresados" class="level2">
<h2 class="anchored" data-anchor-id="genes-diferencialmente-expresados">Genes diferencialmente expresados</h2>
<p>Para selecionar los genes diferecialmente expresados debemos fijar el umbral. En este caso elegimos un FC = 1.5 y un valor de p ajustado de 0.05</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb35"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb35-1"><a href="#cb35-1" aria-hidden="true" tabindex="-1"></a>padj.cutoff <span class="ot">&lt;-</span> <span class="fl">0.05</span></span>
<span id="cb35-2"><a href="#cb35-2" aria-hidden="true" tabindex="-1"></a>lfc.cutoff <span class="ot">&lt;-</span> <span class="fu">log2</span>(<span class="fl">1.5</span>)</span>
<span id="cb35-3"><a href="#cb35-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb35-4"><a href="#cb35-4" aria-hidden="true" tabindex="-1"></a>df.ashr <span class="ot">&lt;-</span> <span class="fu">as.data.frame</span>(resLFC)</span>
<span id="cb35-5"><a href="#cb35-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb35-6"><a href="#cb35-6" aria-hidden="true" tabindex="-1"></a>genes.ashr.up <span class="ot">&lt;-</span> <span class="fu">rownames</span>(df.ashr[df.ashr<span class="sc">$</span>log2FoldChange <span class="sc">&gt;=</span> lfc.cutoff <span class="sc">&amp;</span> df.ashr<span class="sc">$</span>padj <span class="sc">&lt;=</span> padj.cutoff, ])</span>
<span id="cb35-7"><a href="#cb35-7" aria-hidden="true" tabindex="-1"></a>genes.ashr.down <span class="ot">&lt;-</span> <span class="fu">rownames</span>(df.ashr[df.ashr<span class="sc">$</span>log2FoldChange <span class="sc">&lt;=</span> <span class="sc">-</span><span class="dv">1</span> <span class="sc">*</span> lfc.cutoff <span class="sc">&amp;</span> df.ashr<span class="sc">$</span>padj <span class="sc">&lt;=</span> padj.cutoff, ])</span>
<span id="cb35-8"><a href="#cb35-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb35-9"><a href="#cb35-9" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(genes.ashr.up)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "A4GALT"  "ABCA1"   "ABHD14B" "ABL2"    "ACADVL"  "ACAN"   </code></pre>
</div>
<div class="sourceCode cell-code" id="cb37"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb37-1"><a href="#cb37-1" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(genes.ashr.down)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "ABCA9"    "ACTBL2"   "ACTL8"    "ACTR3B"   "ADAMTS16" "ADH1C"   </code></pre>
</div>
</div>
<section id="volcano-plot" class="level3">
<h3 class="anchored" data-anchor-id="volcano-plot">Volcano plot</h3>
<div class="cell">
<div class="sourceCode cell-code" id="cb39"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb39-1"><a href="#cb39-1" aria-hidden="true" tabindex="-1"></a><span class="fu">createVolcanoPlot</span>(df.ashr, lfc.cutoff, padj.cutoff, <span class="at">max.overlaps =</span> <span class="dv">30</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="ejemplo_files/figure-html/volcano-plot-1.png" class="img-fluid figure-img" width="672"></p>
<figcaption>Volcano plot (MOV10_overexpression vs control)</figcaption>
</figure>
</div>
</div>
</div>
</section>
<section id="heatmap" class="level3">
<h3 class="anchored" data-anchor-id="heatmap">Heatmap</h3>
<p>Heatmap de los genes diferecialmente expresados.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb40"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb40-1"><a href="#cb40-1" aria-hidden="true" tabindex="-1"></a>df.norm <span class="ot">&lt;-</span> <span class="fu">counts</span>(dds, <span class="at">normalized =</span> <span class="cn">TRUE</span>)</span>
<span id="cb40-2"><a href="#cb40-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb40-3"><a href="#cb40-3" aria-hidden="true" tabindex="-1"></a>df.dge <span class="ot">&lt;-</span> df.norm[<span class="fu">rownames</span>(df.norm) <span class="sc">%in%</span> <span class="fu">c</span>(genes.ashr.up, genes.ashr.down), ]</span>
<span id="cb40-4"><a href="#cb40-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb40-5"><a href="#cb40-5" aria-hidden="true" tabindex="-1"></a><span class="fu">crearHeatMap</span>(df.dge, metadata)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="ejemplo_files/figure-html/HeatMap-1.png" class="img-fluid figure-img" width="672"></p>
<figcaption>Heatmap</figcaption>
</figure>
</div>
</div>
</div>
</section>
</section>
<section id="guardar-resultados" class="level2">
<h2 class="anchored" data-anchor-id="guardar-resultados">Guardar resultados</h2>
<p>Podemos exportar los datos a ficheros csv</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb41"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb41-1"><a href="#cb41-1" aria-hidden="true" tabindex="-1"></a><span class="fu">write.csv</span>(df.ashr, <span class="st">'data/resultado.csv'</span>, <span class="at">row.names =</span> <span class="cn">TRUE</span>)</span>
<span id="cb41-2"><a href="#cb41-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb41-3"><a href="#cb41-3" aria-hidden="true" tabindex="-1"></a><span class="fu">write.csv</span>(<span class="fu">counts</span>(dds, <span class="at">normalized =</span> <span class="cn">TRUE</span>), </span>
<span id="cb41-4"><a href="#cb41-4" aria-hidden="true" tabindex="-1"></a>                    <span class="st">'data/cuentas_normalizadas.csv'</span>, <span class="at">row.names =</span> <span class="cn">TRUE</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="referencias" class="level2">
<h2 class="anchored" data-anchor-id="referencias">Referencias</h2>
<blockquote class="blockquote">
<p><a href="https://hbctraining.github.io/DGE_workshop/">Harvard Chan Bioinformatics Core (HBC).</a></p>
</blockquote>
<blockquote class="blockquote">
<p>“MOV10 and FMRP regulate AGO2 association with microRNA recognition elements”<br>
(Kenny, P.J. et al, Cell Reports, 11 December 2014).</p>
</blockquote>
<blockquote class="blockquote">
<p><a href="https://bookdown.org/ggiaever/2024_RNA-seq-analysis/">RNA-seq analysis workshop</a></p>
</blockquote>
<blockquote class="blockquote">
<p><a href="https://bioconductor.org/packages/release/bioc/html/DESeq2.html">Differential gene expression analysis based on the negative binomial distribution</a></p>
</blockquote>


</section>

</main> <!-- /main -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const onCopySuccess = function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  }
  const getTextToCopy = function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button:not([data-in-quarto-modal])', {
    text: getTextToCopy
  });
  clipboard.on('success', onCopySuccess);
  if (window.document.getElementById('quarto-embedded-source-code-modal')) {
    // For code content inside modals, clipBoardJS needs to be initialized with a container option
    // TODO: Check when it could be a function (https://github.com/zenorocha/clipboard.js/issues/860)
    const clipboardModal = new window.ClipboardJS('.code-copy-button[data-in-quarto-modal]', {
      text: getTextToCopy,
      container: window.document.getElementById('quarto-embedded-source-code-modal')
    });
    clipboardModal.on('success', onCopySuccess);
  }
    var localhostRegex = new RegExp(/^(?:http|https):\/\/localhost\:?[0-9]*\//);
    var mailtoRegex = new RegExp(/^mailto:/);
      var filterRegex = new RegExp('/' + window.location.host + '/');
    var isInternal = (href) => {
        return filterRegex.test(href) || localhostRegex.test(href) || mailtoRegex.test(href);
    }
    // Inspect non-navigation links and adorn them if external
 	var links = window.document.querySelectorAll('a[href]:not(.nav-link):not(.navbar-brand):not(.toc-action):not(.sidebar-link):not(.sidebar-item-toggle):not(.pagination-link):not(.no-external):not([aria-hidden]):not(.dropdown-item):not(.quarto-navigation-tool):not(.about-link)');
    for (var i=0; i<links.length; i++) {
      const link = links[i];
      if (!isInternal(link.href)) {
        // undo the damage that might have been done by quarto-nav.js in the case of
        // links that we want to consider external
        if (link.dataset.originalHref !== undefined) {
          link.href = link.dataset.originalHref;
        }
      }
    }
  function tippyHover(el, contentFn, onTriggerFn, onUntriggerFn) {
    const config = {
      allowHTML: true,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start',
    };
    if (contentFn) {
      config.content = contentFn;
    }
    if (onTriggerFn) {
      config.onTrigger = onTriggerFn;
    }
    if (onUntriggerFn) {
      config.onUntrigger = onUntriggerFn;
    }
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      if (note) {
        return note.innerHTML;
      } else {
        return "";
      }
    });
  }
  const xrefs = window.document.querySelectorAll('a.quarto-xref');
  const processXRef = (id, note) => {
    // Strip column container classes
    const stripColumnClz = (el) => {
      el.classList.remove("page-full", "page-columns");
      if (el.children) {
        for (const child of el.children) {
          stripColumnClz(child);
        }
      }
    }
    stripColumnClz(note)
    if (id === null || id.startsWith('sec-')) {
      // Special case sections, only their first couple elements
      const container = document.createElement("div");
      if (note.children && note.children.length > 2) {
        container.appendChild(note.children[0].cloneNode(true));
        for (let i = 1; i < note.children.length; i++) {
          const child = note.children[i];
          if (child.tagName === "P" && child.innerText === "") {
            continue;
          } else {
            container.appendChild(child.cloneNode(true));
            break;
          }
        }
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(container);
        }
        return container.innerHTML
      } else {
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(note);
        }
        return note.innerHTML;
      }
    } else {
      // Remove any anchor links if they are present
      const anchorLink = note.querySelector('a.anchorjs-link');
      if (anchorLink) {
        anchorLink.remove();
      }
      if (window.Quarto?.typesetMath) {
        window.Quarto.typesetMath(note);
      }
      // TODO in 1.5, we should make sure this works without a callout special case
      if (note.classList.contains("callout")) {
        return note.outerHTML;
      } else {
        return note.innerHTML;
      }
    }
  }
  for (var i=0; i<xrefs.length; i++) {
    const xref = xrefs[i];
    tippyHover(xref, undefined, function(instance) {
      instance.disable();
      let url = xref.getAttribute('href');
      let hash = undefined; 
      if (url.startsWith('#')) {
        hash = url;
      } else {
        try { hash = new URL(url).hash; } catch {}
      }
      if (hash) {
        const id = hash.replace(/^#\/?/, "");
        const note = window.document.getElementById(id);
        if (note !== null) {
          try {
            const html = processXRef(id, note.cloneNode(true));
            instance.setContent(html);
          } finally {
            instance.enable();
            instance.show();
          }
        } else {
          // See if we can fetch this
          fetch(url.split('#')[0])
          .then(res => res.text())
          .then(html => {
            const parser = new DOMParser();
            const htmlDoc = parser.parseFromString(html, "text/html");
            const note = htmlDoc.getElementById(id);
            if (note !== null) {
              const html = processXRef(id, note);
              instance.setContent(html);
            } 
          }).finally(() => {
            instance.enable();
            instance.show();
          });
        }
      } else {
        // See if we can fetch a full url (with no hash to target)
        // This is a special case and we should probably do some content thinning / targeting
        fetch(url)
        .then(res => res.text())
        .then(html => {
          const parser = new DOMParser();
          const htmlDoc = parser.parseFromString(html, "text/html");
          const note = htmlDoc.querySelector('main.content');
          if (note !== null) {
            // This should only happen for chapter cross references
            // (since there is no id in the URL)
            // remove the first header
            if (note.children.length > 0 && note.children[0].tagName === "HEADER") {
              note.children[0].remove();
            }
            const html = processXRef(null, note);
            instance.setContent(html);
          } 
        }).finally(() => {
          instance.enable();
          instance.show();
        });
      }
    }, function(instance) {
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            div.style.left = 0;
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
        // Handle positioning of the toggle
    window.addEventListener(
      "resize",
      throttle(() => {
        elRect = undefined;
        if (selectedAnnoteEl) {
          selectCodeLines(selectedAnnoteEl);
        }
      }, 10)
    );
    function throttle(fn, ms) {
    let throttle = false;
    let timer;
      return (...args) => {
        if(!throttle) { // first call gets through
            fn.apply(this, args);
            throttle = true;
        } else { // all the others get throttled
            if(timer) clearTimeout(timer); // cancel #2
            timer = setTimeout(() => {
              fn.apply(this, args);
              timer = throttle = false;
            }, ms);
        }
      };
    }
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script>
</div> <!-- /content -->




</body></html>