<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.5.57">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">


<title>ORA (MOV10) – Tumbleweed</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { display: inline-block; text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
</style>


<script src="../site_libs/quarto-nav/quarto-nav.js"></script>
<script src="../site_libs/clipboard/clipboard.min.js"></script>
<script src="../site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="../site_libs/quarto-search/fuse.min.js"></script>
<script src="../site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="../">
<script src="../site_libs/quarto-html/quarto.js"></script>
<script src="../site_libs/quarto-html/popper.min.js"></script>
<script src="../site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="../site_libs/quarto-html/anchor.min.js"></script>
<link href="../site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="../site_libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="../site_libs/bootstrap/bootstrap.min.js"></script>
<link href="../site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="../site_libs/bootstrap/bootstrap.min.css" rel="stylesheet" id="quarto-bootstrap" data-mode="light">
<script id="quarto-search-options" type="application/json">{
  "location": "navbar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "end",
  "type": "overlay",
  "limit": 50,
  "keyboard-shortcut": [
    "f",
    "/",
    "s"
  ],
  "show-item-context": false,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-text-placeholder": "",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit",
    "search-label": "Search"
  }
}</script>


<link rel="stylesheet" href="../styles.css">
</head>

<body class="nav-fixed">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top">
    <nav class="navbar navbar-expand-lg " data-bs-theme="dark">
      <div class="navbar-container container-fluid">
      <div class="navbar-brand-container mx-auto">
    <a class="navbar-brand" href="../index.html">
    <span class="navbar-title">Tumbleweed</span>
    </a>
  </div>
            <div id="quarto-search" class="" title="Search"></div>
          <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbarCollapse" aria-controls="navbarCollapse" role="menu" aria-expanded="false" aria-label="Toggle navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
  <span class="navbar-toggler-icon"></span>
</button>
          <div class="collapse navbar-collapse" id="navbarCollapse">
            <ul class="navbar-nav navbar-nav-scroll me-auto">
  <li class="nav-item dropdown ">
    <a class="nav-link dropdown-toggle" href="#" id="nav-menu-r" role="link" data-bs-toggle="dropdown" aria-expanded="false">
 <span class="menu-text">R</span>
    </a>
    <ul class="dropdown-menu" aria-labelledby="nav-menu-r">    
        <li>
    <a class="dropdown-item" href="../R/introduccion.html">
 <span class="dropdown-text">Introducción a R</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../R/vectores.html">
 <span class="dropdown-text">Vectores</span></a>
  </li>  
    </ul>
  </li>
  <li class="nav-item dropdown ">
    <a class="nav-link dropdown-toggle" href="#" id="nav-menu-rna-seq" role="link" data-bs-toggle="dropdown" aria-expanded="false">
 <span class="menu-text">RNA-seq</span>
    </a>
    <ul class="dropdown-menu" aria-labelledby="nav-menu-rna-seq">    
        <li>
    <a class="dropdown-item" href="../rnaseq/deseq/introduccion.html">
 <span class="dropdown-text">Análisis de expresión diferencial</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../rnaseq/deseq/ejemplo.html">
 <span class="dropdown-text">Caso práctico (MOV10)</span></a>
  </li>  
    </ul>
  </li>
  <li class="nav-item dropdown ">
    <a class="nav-link dropdown-toggle" href="#" id="nav-menu-enriquecimiento-funcional" role="link" data-bs-toggle="dropdown" aria-expanded="false">
 <span class="menu-text">Enriquecimiento funcional</span>
    </a>
    <ul class="dropdown-menu" aria-labelledby="nav-menu-enriquecimiento-funcional">    
        <li>
    <a class="dropdown-item" href="../enrfun/introduccion.html">
 <span class="dropdown-text">Enriquecimiento Funcional</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../enrfun/ora.html">
 <span class="dropdown-text">ORA (MOV10)</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../enrfun/gprofiler.html">
 <span class="dropdown-text">gProfiler (MOV10)</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../enrfun/gsea.html">
 <span class="dropdown-text">GSEA (MOV10)</span></a>
  </li>  
    </ul>
  </li>
  <li class="nav-item dropdown ">
    <a class="nav-link dropdown-toggle" href="#" id="nav-menu-herramientas" role="link" data-bs-toggle="dropdown" aria-expanded="false">
 <span class="menu-text">Herramientas</span>
    </a>
    <ul class="dropdown-menu" aria-labelledby="nav-menu-herramientas">    
        <li>
    <a class="dropdown-item" href="../tools/ensembl.html">
 <span class="dropdown-text">Convertir Ensembl IDs</span></a>
  </li>  
    </ul>
  </li>
  <li class="nav-item dropdown ">
    <a class="nav-link dropdown-toggle" href="#" id="nav-menu-presentaciones" role="link" data-bs-toggle="dropdown" aria-expanded="false">
 <span class="menu-text">Presentaciones</span>
    </a>
    <ul class="dropdown-menu" aria-labelledby="nav-menu-presentaciones">    
        <li>
    <a class="dropdown-item" href="../presentaciones/usaR/usaR.html">
 <span class="dropdown-text">Usa R!</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../presentaciones/ggplot2/ggplot2.html">
 <span class="dropdown-text">ggplot2</span></a>
  </li>  
    </ul>
  </li>
</ul>
          </div> <!-- /navcollapse -->
            <div class="quarto-navbar-tools">
</div>
      </div> <!-- /container-fluid -->
    </nav>
</header>
<!-- content -->
<div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article page-navbar">
<!-- sidebar -->
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        <nav id="TOC" role="doc-toc" class="toc-active">
    <h2 id="toc-title">On this page</h2>
   
  <ul>
  <li><a href="#cargar-datos" id="toc-cargar-datos" class="nav-link active" data-scroll-target="#cargar-datos">Cargar datos</a></li>
  <li><a href="#análisis-go" id="toc-análisis-go" class="nav-link" data-scroll-target="#análisis-go">Análisis GO</a>
  <ul class="collapse">
  <li><a href="#biological-process" id="toc-biological-process" class="nav-link" data-scroll-target="#biological-process">Biological Process</a></li>
  <li><a href="#molecular-function" id="toc-molecular-function" class="nav-link" data-scroll-target="#molecular-function">Molecular Function</a></li>
  <li><a href="#cellular-component" id="toc-cellular-component" class="nav-link" data-scroll-target="#cellular-component">Cellular Component</a></li>
  </ul></li>
  <li><a href="#kegg-pathway-over-representation-analysis" id="toc-kegg-pathway-over-representation-analysis" class="nav-link" data-scroll-target="#kegg-pathway-over-representation-analysis">KEGG pathway over-representation analysis</a></li>
  <li><a href="#pathways" id="toc-pathways" class="nav-link" data-scroll-target="#pathways">Pathways</a></li>
  <li><a href="#referencias" id="toc-referencias" class="nav-link" data-scroll-target="#referencias">Referencias</a></li>
  </ul>
</nav>
    </div>
<!-- main -->
<main class="content" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default">
<div class="quarto-title">
<h1 class="title">ORA (MOV10)</h1>
</div>



<div class="quarto-title-meta">

    
  
    
  </div>
  


</header>


<p>Vamos a realizar un análisis de sobrerepresentación con los genes diferencialmente expresados del dataset MOV10 (<a href="../rnaseq/deseq/ejemplo.html">enlace</a>).</p>
<p>Para realizar el análisis utilizaremos el paquete de Bioconductor <a href="https://bioconductor.org/packages/release/bioc/html/clusterProfiler.html">clusterProfiler</a>, tambien usaremos el paquete <a href="https://bioconductor.org/packages/release/data/annotation/html/org.Hs.eg.db.html">org.Hs.eg.db</a>.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb1"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(clusterProfiler)</span>
<span id="cb1-2"><a href="#cb1-2" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(org.Hs.eg.db)</span>
<span id="cb1-3"><a href="#cb1-3" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(enrichplot)</span>
<span id="cb1-4"><a href="#cb1-4" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(pathview)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<section id="cargar-datos" class="level2">
<h2 class="anchored" data-anchor-id="cargar-datos">Cargar datos</h2>
<p>Cargamos los datos de los genes diferencialmente expresados.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb2"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a>df.fc <span class="ot">&lt;-</span> <span class="fu">read.csv</span>(<span class="st">'../rnaseq/deseq/data/resultado.csv'</span>)</span>
<span id="cb2-2"><a href="#cb2-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-3"><a href="#cb2-3" aria-hidden="true" tabindex="-1"></a>padj.cutoff <span class="ot">&lt;-</span> <span class="fl">0.05</span></span>
<span id="cb2-4"><a href="#cb2-4" aria-hidden="true" tabindex="-1"></a>lfc.cutoff <span class="ot">&lt;-</span> <span class="fu">log2</span>(<span class="fl">1.5</span>)</span>
<span id="cb2-5"><a href="#cb2-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-6"><a href="#cb2-6" aria-hidden="true" tabindex="-1"></a>genes.up <span class="ot">&lt;-</span> df.fc<span class="sc">$</span>X[df.fc<span class="sc">$</span>log2FoldChange <span class="sc">&gt;=</span> lfc.cutoff <span class="sc">&amp;</span> df.fc<span class="sc">$</span>padj <span class="sc">&lt;=</span> padj.cutoff]</span>
<span id="cb2-7"><a href="#cb2-7" aria-hidden="true" tabindex="-1"></a>genes.down <span class="ot">&lt;-</span> df.fc<span class="sc">$</span>X[df.fc<span class="sc">$</span>log2FoldChange <span class="sc">&lt;=</span> <span class="sc">-</span><span class="dv">1</span> <span class="sc">*</span> lfc.cutoff <span class="sc">&amp;</span> df.fc<span class="sc">$</span>padj <span class="sc">&lt;=</span> padj.cutoff]</span>
<span id="cb2-8"><a href="#cb2-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-9"><a href="#cb2-9" aria-hidden="true" tabindex="-1"></a><span class="do">## datos para pathway</span></span>
<span id="cb2-10"><a href="#cb2-10" aria-hidden="true" tabindex="-1"></a>geneList <span class="ot">&lt;-</span> df.fc<span class="sc">$</span>log2FoldChange[df.fc<span class="sc">$</span>X <span class="sc">%in%</span> <span class="fu">c</span>(genes.up, genes.down)]</span>
<span id="cb2-11"><a href="#cb2-11" aria-hidden="true" tabindex="-1"></a><span class="fu">names</span>(geneList) <span class="ot">&lt;-</span> <span class="fu">c</span>(genes.up, genes.down)</span>
<span id="cb2-12"><a href="#cb2-12" aria-hidden="true" tabindex="-1"></a>geneList <span class="ot">&lt;-</span> geneList[<span class="fu">order</span>(geneList, <span class="at">decreasing =</span> <span class="cn">TRUE</span>)]</span>
<span id="cb2-13"><a href="#cb2-13" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-14"><a href="#cb2-14" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-15"><a href="#cb2-15" aria-hidden="true" tabindex="-1"></a>genes.dge <span class="ot">&lt;-</span> <span class="fu">c</span>(genes.up, genes.down)</span>
<span id="cb2-16"><a href="#cb2-16" aria-hidden="true" tabindex="-1"></a>allgenes <span class="ot">&lt;-</span> df.fc<span class="sc">$</span>X</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="análisis-go" class="level2">
<h2 class="anchored" data-anchor-id="análisis-go">Análisis GO</h2>
<p>El paquete clusterProfiler implementa la función <strong>enrichGO()</strong>.</p>
<section id="biological-process" class="level3">
<h3 class="anchored" data-anchor-id="biological-process">Biological Process</h3>
<div class="callout callout-style-default callout-important callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Important
</div>
</div>
<div class="callout-body-container callout-body">
<p>En este ejemplo vamos a utilizar el nombre del gen (<em>SYMBOL</em>) para las consultas ya que es el dato que tenemos en el dataset. Pero esto no es recomendable ya que el nombre del gen puede ser ambiguo. Es mejor utilizar <em>ENSEMBL</em> o <em>ENTREZID</em>. En este caso podemos añadir el parametro <strong><em>readable = TRUE</em></strong> a la función <strong>enrichGO()</strong> para que traduzca los genes ID a nombre de genes.</p>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb3"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb3-1"><a href="#cb3-1" aria-hidden="true" tabindex="-1"></a><span class="fu">keytypes</span>(org.Hs.eg.db)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code> [1] "ACCNUM"       "ALIAS"        "ENSEMBL"      "ENSEMBLPROT"  "ENSEMBLTRANS"
 [6] "ENTREZID"     "ENZYME"       "EVIDENCE"     "EVIDENCEALL"  "GENENAME"    
[11] "GENETYPE"     "GO"           "GOALL"        "IPI"          "MAP"         
[16] "OMIM"         "ONTOLOGY"     "ONTOLOGYALL"  "PATH"         "PFAM"        
[21] "PMID"         "PROSITE"      "REFSEQ"       "SYMBOL"       "UCSCKG"      
[26] "UNIPROT"     </code></pre>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb5"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb5-1"><a href="#cb5-1" aria-hidden="true" tabindex="-1"></a>go.BP <span class="ot">&lt;-</span> <span class="fu">enrichGO</span>(<span class="at">gene          =</span> genes.dge,</span>
<span id="cb5-2"><a href="#cb5-2" aria-hidden="true" tabindex="-1"></a>                  <span class="at">universe      =</span> allgenes,</span>
<span id="cb5-3"><a href="#cb5-3" aria-hidden="true" tabindex="-1"></a>                  <span class="at">OrgDb         =</span> org.Hs.eg.db,</span>
<span id="cb5-4"><a href="#cb5-4" aria-hidden="true" tabindex="-1"></a>                  <span class="at">keyType       =</span> <span class="st">"SYMBOL"</span>,</span>
<span id="cb5-5"><a href="#cb5-5" aria-hidden="true" tabindex="-1"></a>                  <span class="at">ont           =</span> <span class="st">"BP"</span>,</span>
<span id="cb5-6"><a href="#cb5-6" aria-hidden="true" tabindex="-1"></a>                  <span class="at">pAdjustMethod =</span> <span class="st">"BH"</span></span>
<span id="cb5-7"><a href="#cb5-7" aria-hidden="true" tabindex="-1"></a>                )</span>
<span id="cb5-8"><a href="#cb5-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb5-9"><a href="#cb5-9" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(go.BP)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div class="kable-table">
<table class="caption-top table table-sm table-striped small">
<colgroup>
<col style="width: 3%">
<col style="width: 3%">
<col style="width: 12%">
<col style="width: 2%">
<col style="width: 2%">
<col style="width: 3%">
<col style="width: 4%">
<col style="width: 2%">
<col style="width: 2%">
<col style="width: 2%">
<col style="width: 2%">
<col style="width: 55%">
<col style="width: 1%">
</colgroup>
<thead>
<tr class="header">
<th style="text-align: left;"></th>
<th style="text-align: left;">ID</th>
<th style="text-align: left;">Description</th>
<th style="text-align: left;">GeneRatio</th>
<th style="text-align: left;">BgRatio</th>
<th style="text-align: right;">RichFactor</th>
<th style="text-align: right;">FoldEnrichment</th>
<th style="text-align: right;">zScore</th>
<th style="text-align: right;">pvalue</th>
<th style="text-align: right;">p.adjust</th>
<th style="text-align: right;">qvalue</th>
<th style="text-align: left;">geneID</th>
<th style="text-align: right;">Count</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td style="text-align: left;">GO:0030198</td>
<td style="text-align: left;">GO:0030198</td>
<td style="text-align: left;">extracellular matrix organization</td>
<td style="text-align: left;">30/509</td>
<td style="text-align: left;">230/11940</td>
<td style="text-align: right;">0.1304348</td>
<td style="text-align: right;">3.059708</td>
<td style="text-align: right;">6.655671</td>
<td style="text-align: right;">0.00e+00</td>
<td style="text-align: right;">0.0000645</td>
<td style="text-align: right;">0.0000618</td>
<td style="text-align: left;">ADAMTS1/ADAMTS10/ADAMTS2/ADAMTS7/ADAMTSL3/COL13A1/COL23A1/COL2A1/COL3A1/COL5A1/CREB3L1/DPP4/EGFLAM/HAPLN2/ITGA8/LAMA1/LAMB1/LAMC1/LUM/MATN3/MMP2/NID2/NPNT/NTN4/OLFML2A/PAPLN/RECK/SULF1/THSD4/ADAMTS16</td>
<td style="text-align: right;">30</td>
</tr>
<tr class="even">
<td style="text-align: left;">GO:0043062</td>
<td style="text-align: left;">GO:0043062</td>
<td style="text-align: left;">extracellular structure organization</td>
<td style="text-align: left;">30/509</td>
<td style="text-align: left;">231/11940</td>
<td style="text-align: right;">0.1298701</td>
<td style="text-align: right;">3.046462</td>
<td style="text-align: right;">6.627513</td>
<td style="text-align: right;">0.00e+00</td>
<td style="text-align: right;">0.0000645</td>
<td style="text-align: right;">0.0000618</td>
<td style="text-align: left;">ADAMTS1/ADAMTS10/ADAMTS2/ADAMTS7/ADAMTSL3/COL13A1/COL23A1/COL2A1/COL3A1/COL5A1/CREB3L1/DPP4/EGFLAM/HAPLN2/ITGA8/LAMA1/LAMB1/LAMC1/LUM/MATN3/MMP2/NID2/NPNT/NTN4/OLFML2A/PAPLN/RECK/SULF1/THSD4/ADAMTS16</td>
<td style="text-align: right;">30</td>
</tr>
<tr class="odd">
<td style="text-align: left;">GO:0045229</td>
<td style="text-align: left;">GO:0045229</td>
<td style="text-align: left;">external encapsulating structure organization</td>
<td style="text-align: left;">30/509</td>
<td style="text-align: left;">231/11940</td>
<td style="text-align: right;">0.1298701</td>
<td style="text-align: right;">3.046462</td>
<td style="text-align: right;">6.627513</td>
<td style="text-align: right;">0.00e+00</td>
<td style="text-align: right;">0.0000645</td>
<td style="text-align: right;">0.0000618</td>
<td style="text-align: left;">ADAMTS1/ADAMTS10/ADAMTS2/ADAMTS7/ADAMTSL3/COL13A1/COL23A1/COL2A1/COL3A1/COL5A1/CREB3L1/DPP4/EGFLAM/HAPLN2/ITGA8/LAMA1/LAMB1/LAMC1/LUM/MATN3/MMP2/NID2/NPNT/NTN4/OLFML2A/PAPLN/RECK/SULF1/THSD4/ADAMTS16</td>
<td style="text-align: right;">30</td>
</tr>
<tr class="even">
<td style="text-align: left;">GO:0099084</td>
<td style="text-align: left;">GO:0099084</td>
<td style="text-align: left;">postsynaptic specialization organization</td>
<td style="text-align: left;">10/509</td>
<td style="text-align: left;">40/11940</td>
<td style="text-align: right;">0.2500000</td>
<td style="text-align: right;">5.864440</td>
<td style="text-align: right;">6.502646</td>
<td style="text-align: right;">4.80e-06</td>
<td style="text-align: right;">0.0048754</td>
<td style="text-align: right;">0.0046640</td>
<td style="text-align: left;">CNTNAP1/LRRC4B/LRRTM2/NPTX1/NTNG2/SHANK1/SHANK2/LRRC4/NRXN1/ZDHHC12</td>
<td style="text-align: right;">10</td>
</tr>
<tr class="odd">
<td style="text-align: left;">GO:0097106</td>
<td style="text-align: left;">GO:0097106</td>
<td style="text-align: left;">postsynaptic density organization</td>
<td style="text-align: left;">9/509</td>
<td style="text-align: left;">34/11940</td>
<td style="text-align: right;">0.2647059</td>
<td style="text-align: right;">6.209407</td>
<td style="text-align: right;">6.418678</td>
<td style="text-align: right;">8.70e-06</td>
<td style="text-align: right;">0.0070421</td>
<td style="text-align: right;">0.0067368</td>
<td style="text-align: left;">CNTNAP1/LRRC4B/LRRTM2/NPTX1/SHANK1/SHANK2/LRRC4/NRXN1/ZDHHC12</td>
<td style="text-align: right;">9</td>
</tr>
<tr class="even">
<td style="text-align: left;">GO:1901342</td>
<td style="text-align: left;">GO:1901342</td>
<td style="text-align: left;">regulation of vasculature development</td>
<td style="text-align: left;">23/509</td>
<td style="text-align: left;">203/11940</td>
<td style="text-align: right;">0.1133005</td>
<td style="text-align: right;">2.657776</td>
<td style="text-align: right;">5.026851</td>
<td style="text-align: right;">1.81e-05</td>
<td style="text-align: right;">0.0121113</td>
<td style="text-align: right;">0.0115863</td>
<td style="text-align: left;">ADAMTS1/ADM2/ALOX5/CREB3L1/CXCL12/DLL1/EFNA1/FUT1/HSPB1/HSPG2/ITGA5/MDK/PGF/PLK2/PLXND1/RAPGEF3/RECK/RRAS/SULF1/TSPAN18/XBP1/ZC3H12A/ANGPT2</td>
<td style="text-align: right;">23</td>
</tr>
</tbody>
</table>
</div>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb6"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb6-1"><a href="#cb6-1" aria-hidden="true" tabindex="-1"></a><span class="fu">barplot</span>(go.BP, <span class="at">showCategory =</span> <span class="dv">15</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="ora_files/figure-html/barplot-gobp-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
<div class="sourceCode cell-code" id="cb7"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb7-1"><a href="#cb7-1" aria-hidden="true" tabindex="-1"></a><span class="fu">dotplot</span>(go.BP, <span class="at">showCategory =</span> <span class="dv">15</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="ora_files/figure-html/barplot-gobp-2.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
</section>
<section id="molecular-function" class="level3">
<h3 class="anchored" data-anchor-id="molecular-function">Molecular Function</h3>
<div class="cell">
<div class="sourceCode cell-code" id="cb8"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb8-1"><a href="#cb8-1" aria-hidden="true" tabindex="-1"></a>go.MF <span class="ot">&lt;-</span> <span class="fu">enrichGO</span>(<span class="at">gene          =</span> genes.dge,</span>
<span id="cb8-2"><a href="#cb8-2" aria-hidden="true" tabindex="-1"></a>                  <span class="at">universe      =</span> allgenes,</span>
<span id="cb8-3"><a href="#cb8-3" aria-hidden="true" tabindex="-1"></a>                  <span class="at">OrgDb         =</span> org.Hs.eg.db,</span>
<span id="cb8-4"><a href="#cb8-4" aria-hidden="true" tabindex="-1"></a>                  <span class="at">keyType       =</span> <span class="st">"SYMBOL"</span>,</span>
<span id="cb8-5"><a href="#cb8-5" aria-hidden="true" tabindex="-1"></a>                  <span class="at">ont           =</span> <span class="st">"MF"</span>,</span>
<span id="cb8-6"><a href="#cb8-6" aria-hidden="true" tabindex="-1"></a>                  <span class="at">pAdjustMethod =</span> <span class="st">"BH"</span></span>
<span id="cb8-7"><a href="#cb8-7" aria-hidden="true" tabindex="-1"></a>                )</span>
<span id="cb8-8"><a href="#cb8-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb8-9"><a href="#cb8-9" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(go.MF)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div class="kable-table">
<table class="caption-top table table-sm table-striped small">
<colgroup>
<col style="width: 2%">
<col style="width: 2%">
<col style="width: 12%">
<col style="width: 2%">
<col style="width: 2%">
<col style="width: 2%">
<col style="width: 3%">
<col style="width: 2%">
<col style="width: 2%">
<col style="width: 2%">
<col style="width: 2%">
<col style="width: 57%">
<col style="width: 1%">
</colgroup>
<thead>
<tr class="header">
<th style="text-align: left;"></th>
<th style="text-align: left;">ID</th>
<th style="text-align: left;">Description</th>
<th style="text-align: left;">GeneRatio</th>
<th style="text-align: left;">BgRatio</th>
<th style="text-align: right;">RichFactor</th>
<th style="text-align: right;">FoldEnrichment</th>
<th style="text-align: right;">zScore</th>
<th style="text-align: right;">pvalue</th>
<th style="text-align: right;">p.adjust</th>
<th style="text-align: right;">qvalue</th>
<th style="text-align: left;">geneID</th>
<th style="text-align: right;">Count</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td style="text-align: left;">GO:0005201</td>
<td style="text-align: left;">GO:0005201</td>
<td style="text-align: left;">extracellular matrix structural constituent</td>
<td style="text-align: left;">24/506</td>
<td style="text-align: left;">105/12095</td>
<td style="text-align: right;">0.2285714</td>
<td style="text-align: right;">5.463580</td>
<td style="text-align: right;">9.598550</td>
<td style="text-align: right;">0.0000000</td>
<td style="text-align: right;">0.0000000</td>
<td style="text-align: right;">0.0000000</td>
<td style="text-align: left;">ACAN/AGRN/COL13A1/COL21A1/COL23A1/COL2A1/COL3A1/COL5A1/COL6A2/EFEMP1/FBN3/HSPG2/IGFBP7/LAMA1/LAMB1/LAMC1/LTBP4/LUM/MATN3/NID2/NPNT/SRPX/THBS3/THSD4</td>
<td style="text-align: right;">24</td>
</tr>
<tr class="even">
<td style="text-align: left;">GO:0005539</td>
<td style="text-align: left;">GO:0005539</td>
<td style="text-align: left;">glycosaminoglycan binding</td>
<td style="text-align: left;">19/506</td>
<td style="text-align: left;">132/12095</td>
<td style="text-align: right;">0.1439394</td>
<td style="text-align: right;">3.440607</td>
<td style="text-align: right;">5.891176</td>
<td style="text-align: right;">0.0000023</td>
<td style="text-align: right;">0.0007422</td>
<td style="text-align: right;">0.0007050</td>
<td style="text-align: left;">ACAN/ADAMTS1/AGRN/APLP2/CD44/COL13A1/COL23A1/COL5A1/EGFLAM/HAPLN2/LTBP4/MDK/NOD2/PGF/SLIT3/SULF1/THBS3/HDGF/TMEM184A</td>
<td style="text-align: right;">19</td>
</tr>
<tr class="odd">
<td style="text-align: left;">GO:0005509</td>
<td style="text-align: left;">GO:0005509</td>
<td style="text-align: left;">calcium ion binding</td>
<td style="text-align: left;">36/506</td>
<td style="text-align: left;">445/12095</td>
<td style="text-align: right;">0.0808989</td>
<td style="text-align: right;">1.933739</td>
<td style="text-align: right;">4.193530</td>
<td style="text-align: right;">0.0001146</td>
<td style="text-align: right;">0.0242589</td>
<td style="text-align: right;">0.0230425</td>
<td style="text-align: left;">AGRN/CAPN12/CDH23/CDH3/CDHR1/DCHS1/DLL1/DYSF/EFEMP1/EGFLAM/FAM20C/FBN3/GALNT3/HSPG2/ITGA5/ITPR3/LTBP4/MATN3/MEGF6/NID2/NOTCH3/NPNT/PCDH18/PCDHB10/RASGRP2/SCUBE1/SCUBE3/SLIT3/SNED1/SPOCK1/SULF1/SYT17/SYT2/SYT6/THBS3/NRXN1</td>
<td style="text-align: right;">36</td>
</tr>
<tr class="even">
<td style="text-align: left;">GO:0030546</td>
<td style="text-align: left;">GO:0030546</td>
<td style="text-align: left;">signaling receptor activator activity</td>
<td style="text-align: left;">22/506</td>
<td style="text-align: left;">224/12095</td>
<td style="text-align: right;">0.0982143</td>
<td style="text-align: right;">2.347632</td>
<td style="text-align: right;">4.253916</td>
<td style="text-align: right;">0.0001774</td>
<td style="text-align: right;">0.0281657</td>
<td style="text-align: right;">0.0267533</td>
<td style="text-align: left;">ADM2/CHGB/CRLF1/CXCL12/DLL1/DPP4/EFEMP1/EGFR/GDF15/GDNF/HSPA1A/IL11/IL32/INHBE/MDK/METRNL/NRTN/PGF/SCG2/SEMA5B/TSLP/HDGF</td>
<td style="text-align: right;">22</td>
</tr>
<tr class="odd">
<td style="text-align: left;">GO:0048018</td>
<td style="text-align: left;">GO:0048018</td>
<td style="text-align: left;">receptor ligand activity</td>
<td style="text-align: left;">21/506</td>
<td style="text-align: left;">219/12095</td>
<td style="text-align: right;">0.0958904</td>
<td style="text-align: right;">2.292084</td>
<td style="text-align: right;">4.031949</td>
<td style="text-align: right;">0.0003432</td>
<td style="text-align: right;">0.0435811</td>
<td style="text-align: right;">0.0413957</td>
<td style="text-align: left;">ADM2/CHGB/CRLF1/CXCL12/DLL1/DPP4/EFEMP1/GDF15/GDNF/HSPA1A/IL11/IL32/INHBE/MDK/METRNL/NRTN/PGF/SCG2/SEMA5B/TSLP/HDGF</td>
<td style="text-align: right;">21</td>
</tr>
<tr class="even">
<td style="text-align: left;">GO:0022843</td>
<td style="text-align: left;">GO:0022843</td>
<td style="text-align: left;">voltage-gated monoatomic cation channel activity</td>
<td style="text-align: left;">11/506</td>
<td style="text-align: left;">79/12095</td>
<td style="text-align: right;">0.1392405</td>
<td style="text-align: right;">3.328288</td>
<td style="text-align: right;">4.338180</td>
<td style="text-align: right;">0.0004228</td>
<td style="text-align: right;">0.0440472</td>
<td style="text-align: right;">0.0418384</td>
<td style="text-align: left;">CACNA1G/CACNG6/HVCN1/KCNAB2/KCNG1/KCNH3/KCNJ4/KCNQ1/KCNT2/SNAP25/TRPV1</td>
<td style="text-align: right;">11</td>
</tr>
</tbody>
</table>
</div>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb9"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb9-1"><a href="#cb9-1" aria-hidden="true" tabindex="-1"></a><span class="fu">barplot</span>(go.MF, <span class="at">showCategory =</span> <span class="dv">15</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="ora_files/figure-html/barplot-gomf-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
<div class="sourceCode cell-code" id="cb10"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb10-1"><a href="#cb10-1" aria-hidden="true" tabindex="-1"></a><span class="fu">dotplot</span>(go.MF, <span class="at">showCategory =</span> <span class="dv">15</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="ora_files/figure-html/barplot-gomf-2.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
</section>
<section id="cellular-component" class="level3">
<h3 class="anchored" data-anchor-id="cellular-component">Cellular Component</h3>
<div class="cell">
<div class="sourceCode cell-code" id="cb11"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb11-1"><a href="#cb11-1" aria-hidden="true" tabindex="-1"></a>go.CC <span class="ot">&lt;-</span> <span class="fu">enrichGO</span>(<span class="at">gene          =</span> genes.dge,</span>
<span id="cb11-2"><a href="#cb11-2" aria-hidden="true" tabindex="-1"></a>                  <span class="at">universe      =</span> allgenes,</span>
<span id="cb11-3"><a href="#cb11-3" aria-hidden="true" tabindex="-1"></a>                  <span class="at">OrgDb         =</span> org.Hs.eg.db,</span>
<span id="cb11-4"><a href="#cb11-4" aria-hidden="true" tabindex="-1"></a>                  <span class="at">keyType       =</span> <span class="st">"SYMBOL"</span>,</span>
<span id="cb11-5"><a href="#cb11-5" aria-hidden="true" tabindex="-1"></a>                  <span class="at">ont           =</span> <span class="st">"CC"</span>,</span>
<span id="cb11-6"><a href="#cb11-6" aria-hidden="true" tabindex="-1"></a>                  <span class="at">pAdjustMethod =</span> <span class="st">"BH"</span></span>
<span id="cb11-7"><a href="#cb11-7" aria-hidden="true" tabindex="-1"></a>                )</span>
<span id="cb11-8"><a href="#cb11-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb11-9"><a href="#cb11-9" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(go.CC)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div class="kable-table">
<table class="caption-top table table-sm table-striped small">
<colgroup>
<col style="width: 2%">
<col style="width: 2%">
<col style="width: 9%">
<col style="width: 2%">
<col style="width: 2%">
<col style="width: 2%">
<col style="width: 3%">
<col style="width: 1%">
<col style="width: 2%">
<col style="width: 2%">
<col style="width: 2%">
<col style="width: 66%">
<col style="width: 1%">
</colgroup>
<thead>
<tr class="header">
<th style="text-align: left;"></th>
<th style="text-align: left;">ID</th>
<th style="text-align: left;">Description</th>
<th style="text-align: left;">GeneRatio</th>
<th style="text-align: left;">BgRatio</th>
<th style="text-align: right;">RichFactor</th>
<th style="text-align: right;">FoldEnrichment</th>
<th style="text-align: right;">zScore</th>
<th style="text-align: right;">pvalue</th>
<th style="text-align: right;">p.adjust</th>
<th style="text-align: right;">qvalue</th>
<th style="text-align: left;">geneID</th>
<th style="text-align: right;">Count</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td style="text-align: left;">GO:0062023</td>
<td style="text-align: left;">GO:0062023</td>
<td style="text-align: left;">collagen-containing extracellular matrix</td>
<td style="text-align: left;">42/528</td>
<td style="text-align: left;">259/12370</td>
<td style="text-align: right;">0.1621622</td>
<td style="text-align: right;">3.799140</td>
<td style="text-align: right;">9.612940</td>
<td style="text-align: right;">0.0000000</td>
<td style="text-align: right;">0.0000000</td>
<td style="text-align: right;">0.0000000</td>
<td style="text-align: left;">ACAN/ACHE/ADAMTS1/ADAMTS10/ADAMTS2/AGRN/BCAM/COL13A1/COL21A1/COL23A1/COL2A1/COL3A1/COL5A1/COL6A2/CXCL12/EFEMP1/EGFLAM/GDF15/HAPLN2/HSPG2/HTRA1/IGFBP7/INHBE/ITIH5/LAMA1/LAMB1/LAMC1/LTBP4/LUM/MATN3/MDK/MMP2/NID2/NPNT/NTN4/SRPX/SULF1/THBS3/THSD4/ANGPT2/HDGF/HNRNPM</td>
<td style="text-align: right;">42</td>
</tr>
<tr class="even">
<td style="text-align: left;">GO:0030312</td>
<td style="text-align: left;">GO:0030312</td>
<td style="text-align: left;">external encapsulating structure</td>
<td style="text-align: left;">48/528</td>
<td style="text-align: left;">335/12370</td>
<td style="text-align: right;">0.1432836</td>
<td style="text-align: right;">3.356852</td>
<td style="text-align: right;">9.234288</td>
<td style="text-align: right;">0.0000000</td>
<td style="text-align: right;">0.0000000</td>
<td style="text-align: right;">0.0000000</td>
<td style="text-align: left;">ACAN/ACHE/ADAMTS1/ADAMTS10/ADAMTS2/ADAMTS7/ADAMTSL3/AGRN/BCAM/COL13A1/COL21A1/COL23A1/COL2A1/COL3A1/COL5A1/COL6A2/CXCL12/EFEMP1/EGFLAM/FBN3/GDF15/HAPLN2/HSPG2/HTRA1/IGFBP7/INHBE/ITIH5/LAMA1/LAMB1/LAMC1/LTBP4/LUM/MATN3/MDK/MMP2/NID2/NPNT/NTN4/OLFML2A/SNED1/SRPX/SULF1/THBS3/THSD4/ADAMTS16/ANGPT2/HDGF/HNRNPM</td>
<td style="text-align: right;">48</td>
</tr>
<tr class="odd">
<td style="text-align: left;">GO:0031012</td>
<td style="text-align: left;">GO:0031012</td>
<td style="text-align: left;">extracellular matrix</td>
<td style="text-align: left;">48/528</td>
<td style="text-align: left;">335/12370</td>
<td style="text-align: right;">0.1432836</td>
<td style="text-align: right;">3.356852</td>
<td style="text-align: right;">9.234288</td>
<td style="text-align: right;">0.0000000</td>
<td style="text-align: right;">0.0000000</td>
<td style="text-align: right;">0.0000000</td>
<td style="text-align: left;">ACAN/ACHE/ADAMTS1/ADAMTS10/ADAMTS2/ADAMTS7/ADAMTSL3/AGRN/BCAM/COL13A1/COL21A1/COL23A1/COL2A1/COL3A1/COL5A1/COL6A2/CXCL12/EFEMP1/EGFLAM/FBN3/GDF15/HAPLN2/HSPG2/HTRA1/IGFBP7/INHBE/ITIH5/LAMA1/LAMB1/LAMC1/LTBP4/LUM/MATN3/MDK/MMP2/NID2/NPNT/NTN4/OLFML2A/SNED1/SRPX/SULF1/THBS3/THSD4/ADAMTS16/ANGPT2/HDGF/HNRNPM</td>
<td style="text-align: right;">48</td>
</tr>
<tr class="even">
<td style="text-align: left;">GO:0005604</td>
<td style="text-align: left;">GO:0005604</td>
<td style="text-align: left;">basement membrane</td>
<td style="text-align: left;">14/528</td>
<td style="text-align: left;">76/12370</td>
<td style="text-align: right;">0.1842105</td>
<td style="text-align: right;">4.315690</td>
<td style="text-align: right;">6.122176</td>
<td style="text-align: right;">0.0000034</td>
<td style="text-align: right;">0.0003562</td>
<td style="text-align: right;">0.0003404</td>
<td style="text-align: left;">ACHE/ADAMTS1/AGRN/COL23A1/COL2A1/COL5A1/EGFLAM/HSPG2/LAMA1/LAMB1/LAMC1/NID2/NPNT/NTN4</td>
<td style="text-align: right;">14</td>
</tr>
<tr class="odd">
<td style="text-align: left;">GO:0098636</td>
<td style="text-align: left;">GO:0098636</td>
<td style="text-align: left;">protein complex involved in cell adhesion</td>
<td style="text-align: left;">9/528</td>
<td style="text-align: left;">40/12370</td>
<td style="text-align: right;">0.2250000</td>
<td style="text-align: right;">5.271307</td>
<td style="text-align: right;">5.713220</td>
<td style="text-align: right;">0.0000367</td>
<td style="text-align: right;">0.0031137</td>
<td style="text-align: right;">0.0029761</td>
<td style="text-align: left;">ITGA11/ITGA4/ITGA5/ITGA8/JAM2/LAMA1/LAMB1/LAMC1/PLAU</td>
<td style="text-align: right;">9</td>
</tr>
<tr class="even">
<td style="text-align: left;">GO:0009897</td>
<td style="text-align: left;">GO:0009897</td>
<td style="text-align: left;">external side of plasma membrane</td>
<td style="text-align: left;">21/528</td>
<td style="text-align: left;">200/12370</td>
<td style="text-align: right;">0.1050000</td>
<td style="text-align: right;">2.459943</td>
<td style="text-align: right;">4.395187</td>
<td style="text-align: right;">0.0001286</td>
<td style="text-align: right;">0.0090870</td>
<td style="text-align: right;">0.0086854</td>
<td style="text-align: left;">ABCA1/ACE/ATP1B2/BCAM/CD24/CRLF1/CXCL12/FCGRT/IL1R1/ITGA11/ITGA4/ITGA5/ITGA8/MCAM/PLAU/SCUBE1/TNFRSF13C/TNFRSF9/ULBP2/CCR6/TRPV1</td>
<td style="text-align: right;">21</td>
</tr>
</tbody>
</table>
</div>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb12"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb12-1"><a href="#cb12-1" aria-hidden="true" tabindex="-1"></a><span class="fu">barplot</span>(go.CC, <span class="at">showCategory =</span> <span class="dv">15</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="ora_files/figure-html/barplot-gocc-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
<div class="sourceCode cell-code" id="cb13"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb13-1"><a href="#cb13-1" aria-hidden="true" tabindex="-1"></a><span class="fu">dotplot</span>(go.CC, <span class="at">showCategory =</span> <span class="dv">15</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="ora_files/figure-html/barplot-gocc-2.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
</section>
</section>
<section id="kegg-pathway-over-representation-analysis" class="level2">
<h2 class="anchored" data-anchor-id="kegg-pathway-over-representation-analysis">KEGG pathway over-representation analysis</h2>
<p>El paquete <strong>clusterProfiler</strong> soporta todos los organismos que tienen anotación de la base da datos <a href="https://www.kegg.jp/">KEGG</a>. Para realizar el análisis de sobrerepresentación usaremos la función <strong>enrichKEGG()</strong>. Esta función no acepta los genes con nombre (<em>SYMBOL</em>), el valor debe de ser <em>ENTREZID</em>. Podemos traducir los simbolos usando la función <strong>bitr()</strong>(Biological Id TranslatoR).</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb14"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb14-1"><a href="#cb14-1" aria-hidden="true" tabindex="-1"></a>allgenes.entrezid <span class="ot">&lt;-</span> <span class="fu">bitr</span>(allgenes, </span>
<span id="cb14-2"><a href="#cb14-2" aria-hidden="true" tabindex="-1"></a>                          <span class="at">fromType =</span> <span class="st">'SYMBOL'</span>,</span>
<span id="cb14-3"><a href="#cb14-3" aria-hidden="true" tabindex="-1"></a>                          <span class="at">toType =</span> <span class="st">'ENTREZID'</span>,</span>
<span id="cb14-4"><a href="#cb14-4" aria-hidden="true" tabindex="-1"></a>                          <span class="at">OrgDb =</span> <span class="st">'org.Hs.eg.db'</span></span>
<span id="cb14-5"><a href="#cb14-5" aria-hidden="true" tabindex="-1"></a>                          )</span>
<span id="cb14-6"><a href="#cb14-6" aria-hidden="true" tabindex="-1"></a>genes.dge.entrezid <span class="ot">&lt;-</span> <span class="fu">bitr</span>(genes.dge, </span>
<span id="cb14-7"><a href="#cb14-7" aria-hidden="true" tabindex="-1"></a>                          <span class="at">fromType =</span> <span class="st">'SYMBOL'</span>,</span>
<span id="cb14-8"><a href="#cb14-8" aria-hidden="true" tabindex="-1"></a>                          <span class="at">toType =</span> <span class="st">'ENTREZID'</span>,</span>
<span id="cb14-9"><a href="#cb14-9" aria-hidden="true" tabindex="-1"></a>                          <span class="at">OrgDb =</span> <span class="st">'org.Hs.eg.db'</span></span>
<span id="cb14-10"><a href="#cb14-10" aria-hidden="true" tabindex="-1"></a>                          )</span>
<span id="cb14-11"><a href="#cb14-11" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(genes.dge.entrezid)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div class="kable-table">
<table class="caption-top table table-sm table-striped small">
<thead>
<tr class="header">
<th style="text-align: left;">SYMBOL</th>
<th style="text-align: left;">ENTREZID</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td style="text-align: left;">A4GALT</td>
<td style="text-align: left;">53947</td>
</tr>
<tr class="even">
<td style="text-align: left;">ABCA1</td>
<td style="text-align: left;">19</td>
</tr>
<tr class="odd">
<td style="text-align: left;">ABHD14B</td>
<td style="text-align: left;">84836</td>
</tr>
<tr class="even">
<td style="text-align: left;">ABL2</td>
<td style="text-align: left;">27</td>
</tr>
<tr class="odd">
<td style="text-align: left;">ACADVL</td>
<td style="text-align: left;">37</td>
</tr>
<tr class="even">
<td style="text-align: left;">ACAN</td>
<td style="text-align: left;">176</td>
</tr>
</tbody>
</table>
</div>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb15"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb15-1"><a href="#cb15-1" aria-hidden="true" tabindex="-1"></a>kegg <span class="ot">&lt;-</span> <span class="fu">enrichKEGG</span>(<span class="at">gene         =</span> genes.dge.entrezid<span class="sc">$</span>ENTREZID,</span>
<span id="cb15-2"><a href="#cb15-2" aria-hidden="true" tabindex="-1"></a>                   <span class="at">organism     =</span> <span class="st">'hsa'</span>,</span>
<span id="cb15-3"><a href="#cb15-3" aria-hidden="true" tabindex="-1"></a>                   <span class="at">universe     =</span> allgenes.entrezid<span class="sc">$</span>ENTREZID)</span>
<span id="cb15-4"><a href="#cb15-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-5"><a href="#cb15-5" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(kegg)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div class="kable-table">
<table class="caption-top table table-sm table-striped small">
<colgroup>
<col style="width: 2%">
<col style="width: 10%">
<col style="width: 10%">
<col style="width: 2%">
<col style="width: 9%">
<col style="width: 2%">
<col style="width: 2%">
<col style="width: 3%">
<col style="width: 4%">
<col style="width: 2%">
<col style="width: 2%">
<col style="width: 2%">
<col style="width: 2%">
<col style="width: 38%">
<col style="width: 1%">
</colgroup>
<thead>
<tr class="header">
<th style="text-align: left;"></th>
<th style="text-align: left;">category</th>
<th style="text-align: left;">subcategory</th>
<th style="text-align: left;">ID</th>
<th style="text-align: left;">Description</th>
<th style="text-align: left;">GeneRatio</th>
<th style="text-align: left;">BgRatio</th>
<th style="text-align: right;">RichFactor</th>
<th style="text-align: right;">FoldEnrichment</th>
<th style="text-align: right;">zScore</th>
<th style="text-align: right;">pvalue</th>
<th style="text-align: right;">p.adjust</th>
<th style="text-align: right;">qvalue</th>
<th style="text-align: left;">geneID</th>
<th style="text-align: right;">Count</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td style="text-align: left;">hsa04512</td>
<td style="text-align: left;">Environmental Information Processing</td>
<td style="text-align: left;">Signaling molecules and interaction</td>
<td style="text-align: left;">hsa04512</td>
<td style="text-align: left;">ECM-receptor interaction</td>
<td style="text-align: left;">14/271</td>
<td style="text-align: left;">70/5880</td>
<td style="text-align: right;">0.2000000</td>
<td style="text-align: right;">4.339483</td>
<td style="text-align: right;">6.177801</td>
<td style="text-align: right;">0.0000026</td>
<td style="text-align: right;">0.0007670</td>
<td style="text-align: right;">0.0007380</td>
<td style="text-align: left;">375790/960/1280/1292/3339/22801/3676/3678/8516/284217/3912/3915/255743/7059</td>
<td style="text-align: right;">14</td>
</tr>
<tr class="even">
<td style="text-align: left;">hsa04820</td>
<td style="text-align: left;">NA</td>
<td style="text-align: left;">NA</td>
<td style="text-align: left;">hsa04820</td>
<td style="text-align: left;">Cytoskeleton in muscle cells</td>
<td style="text-align: left;">20/271</td>
<td style="text-align: left;">176/5880</td>
<td style="text-align: right;">0.1136364</td>
<td style="text-align: right;">2.465616</td>
<td style="text-align: right;">4.338907</td>
<td style="text-align: right;">0.0001513</td>
<td style="text-align: right;">0.0220196</td>
<td style="text-align: right;">0.0211873</td>
<td style="text-align: left;">375790/482/1281/1289/1292/2027/84467/2318/3339/22801/3676/3678/8516/284217/25802/22795/84033/7059/22989/4703</td>
<td style="text-align: right;">20</td>
</tr>
<tr class="odd">
<td style="text-align: left;">hsa04151</td>
<td style="text-align: left;">Environmental Information Processing</td>
<td style="text-align: left;">Signal transduction</td>
<td style="text-align: left;">hsa04151</td>
<td style="text-align: left;">PI3K-Akt signaling pathway</td>
<td style="text-align: left;">26/271</td>
<td style="text-align: left;">276/5880</td>
<td style="text-align: right;">0.0942029</td>
<td style="text-align: right;">2.043960</td>
<td style="text-align: right;">3.904656</td>
<td style="text-align: right;">0.0003581</td>
<td style="text-align: right;">0.0298425</td>
<td style="text-align: right;">0.0287145</td>
<td style="text-align: left;">595/1026/1280/1292/90993/64764/54541/1942/1956/2065/2668/22801/3676/3678/8516/3718/284217/3912/3915/9863/4902/5105/5228/7059/285/2790</td>
<td style="text-align: right;">26</td>
</tr>
<tr class="even">
<td style="text-align: left;">hsa04974</td>
<td style="text-align: left;">Organismal Systems</td>
<td style="text-align: left;">Digestive system</td>
<td style="text-align: left;">hsa04974</td>
<td style="text-align: left;">Protein digestion and absorption</td>
<td style="text-align: left;">10/271</td>
<td style="text-align: left;">61/5880</td>
<td style="text-align: right;">0.1639344</td>
<td style="text-align: right;">3.556954</td>
<td style="text-align: right;">4.412221</td>
<td style="text-align: right;">0.0004102</td>
<td style="text-align: right;">0.0298425</td>
<td style="text-align: right;">0.0287145</td>
<td style="text-align: left;">482/1305/81578/91522/1280/1281/1289/1292/1803/3784</td>
<td style="text-align: right;">10</td>
</tr>
</tbody>
</table>
</div>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb16"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb16-1"><a href="#cb16-1" aria-hidden="true" tabindex="-1"></a><span class="fu">dotplot</span>(kegg, <span class="at">showCategory =</span> <span class="dv">15</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="ora_files/figure-html/dotplot-kegg-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p>En el paquete <strong>clusterProfiler</strong> viene incluida la funcion <strong>browseKEGG()</strong> que ayuda a explorar los vías enriquecidas de KEGG con los genes de interes. La función <strong>browseKEGG()</strong> abrira la ruta especificada en un navegador destacando los genes de interes.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb17"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb17-1"><a href="#cb17-1" aria-hidden="true" tabindex="-1"></a><span class="fu">browseKEGG</span>(kegg, <span class="st">'hsa04512'</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="../assets/hsa04512_20241103_195314.png" class="img-fluid quarto-figure quarto-figure-center figure-img" style="width:70.0%"></p>
</figure>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb18"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb18-1"><a href="#cb18-1" aria-hidden="true" tabindex="-1"></a><span class="fu">browseKEGG</span>(kegg, <span class="st">'hsa04151'</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="../assets/hsa04151_20241103_195304.png" class="img-fluid quarto-figure quarto-figure-center figure-img" style="width:70.0%"></p>
</figure>
</div>
</section>
<section id="pathways" class="level2">
<h2 class="anchored" data-anchor-id="pathways">Pathways</h2>
<div class="cell">
<div class="sourceCode cell-code" id="cb19"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb19-1"><a href="#cb19-1" aria-hidden="true" tabindex="-1"></a>genes.dge.entrezid <span class="ot">&lt;-</span> <span class="fu">bitr</span>(<span class="fu">names</span>(geneList), </span>
<span id="cb19-2"><a href="#cb19-2" aria-hidden="true" tabindex="-1"></a>                          <span class="at">fromType =</span> <span class="st">'SYMBOL'</span>,</span>
<span id="cb19-3"><a href="#cb19-3" aria-hidden="true" tabindex="-1"></a>                          <span class="at">toType =</span> <span class="st">'ENTREZID'</span>,</span>
<span id="cb19-4"><a href="#cb19-4" aria-hidden="true" tabindex="-1"></a>                          <span class="at">OrgDb =</span> <span class="st">'org.Hs.eg.db'</span></span>
<span id="cb19-5"><a href="#cb19-5" aria-hidden="true" tabindex="-1"></a>                          )</span>
<span id="cb19-6"><a href="#cb19-6" aria-hidden="true" tabindex="-1"></a><span class="fu">names</span>(geneList) <span class="ot">&lt;-</span> genes.dge.entrezid<span class="sc">$</span>ENTREZID</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb20"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb20-1"><a href="#cb20-1" aria-hidden="true" tabindex="-1"></a>core <span class="ot">&lt;-</span> <span class="fu">geneInCategory</span>(kegg)[[<span class="st">'hsa04512'</span>]]</span>
<span id="cb20-2"><a href="#cb20-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb20-3"><a href="#cb20-3" aria-hidden="true" tabindex="-1"></a>hsa04512 <span class="ot">&lt;-</span> <span class="fu">pathview</span>(<span class="at">gene.data  =</span> geneList,</span>
<span id="cb20-4"><a href="#cb20-4" aria-hidden="true" tabindex="-1"></a>                     <span class="at">pathway.id =</span> <span class="st">"hsa04512"</span>,</span>
<span id="cb20-5"><a href="#cb20-5" aria-hidden="true" tabindex="-1"></a>                     <span class="at">species    =</span> <span class="st">"hsa"</span>,</span>
<span id="cb20-6"><a href="#cb20-6" aria-hidden="true" tabindex="-1"></a>                     <span class="at">limit      =</span> <span class="fu">list</span>(<span class="at">gene=</span><span class="fu">round</span>(<span class="fu">max</span>(<span class="fu">abs</span>(geneList[core])), <span class="dv">1</span>), <span class="at">cpd=</span><span class="dv">1</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="hsa04512.pathview.png" class="img-fluid quarto-figure quarto-figure-center figure-img" style="width:70.0%"></p>
</figure>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb21"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb21-1"><a href="#cb21-1" aria-hidden="true" tabindex="-1"></a>core <span class="ot">&lt;-</span> <span class="fu">geneInCategory</span>(kegg)[[<span class="st">'hsa04151'</span>]]</span>
<span id="cb21-2"><a href="#cb21-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb21-3"><a href="#cb21-3" aria-hidden="true" tabindex="-1"></a>hsa04151<span class="ot">&lt;-</span> <span class="fu">pathview</span>(<span class="at">gene.data  =</span> geneList,</span>
<span id="cb21-4"><a href="#cb21-4" aria-hidden="true" tabindex="-1"></a>                     <span class="at">pathway.id =</span> <span class="st">"hsa04151"</span>,</span>
<span id="cb21-5"><a href="#cb21-5" aria-hidden="true" tabindex="-1"></a>                     <span class="at">species    =</span> <span class="st">"hsa"</span>,</span>
<span id="cb21-6"><a href="#cb21-6" aria-hidden="true" tabindex="-1"></a>                     <span class="at">limit      =</span> <span class="fu">list</span>(<span class="at">gene=</span><span class="fu">round</span>(<span class="fu">max</span>(<span class="fu">abs</span>(geneList[core])), <span class="dv">1</span>), <span class="at">cpd=</span><span class="dv">1</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="hsa04151.pathview.png" class="img-fluid quarto-figure quarto-figure-center figure-img" style="width:70.0%"></p>
</figure>
</div>
</section>
<section id="referencias" class="level2">
<h2 class="anchored" data-anchor-id="referencias">Referencias</h2>
<div class="cell">
<div class="sourceCode cell-code" id="cb22"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb22-1"><a href="#cb22-1" aria-hidden="true" tabindex="-1"></a><span class="fu">citation</span>(<span class="st">"clusterProfiler"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Please cite S. Xu (2024) for using clusterProfiler. In addition, please
cite G. Yu (2010) when using GOSemSim, G. Yu (2015) when using DOSE and
G. Yu (2015) when using ChIPseeker.

  S Xu, E Hu, Y Cai, Z Xie, X Luo, L Zhan, W Tang, Q Wang, B Liu, R
  Wang, W Xie, T Wu, L Xie, G Yu. Using clusterProfiler to characterize
  multiomics data. Nature Protocols. 2024,
  doi:10.1038/s41596-024-01020-z

  T Wu, E Hu, S Xu, M Chen, P Guo, Z Dai, T Feng, L Zhou, W Tang, L
  Zhan, X Fu, S Liu, X Bo, and G Yu. clusterProfiler 4.0: A universal
  enrichment tool for interpreting omics data. The Innovation. 2021,
  2(3):100141

  Guangchuang Yu, Li-Gen Wang, Yanyan Han and Qing-Yu He.
  clusterProfiler: an R package for comparing biological themes among
  gene clusters. OMICS: A Journal of Integrative Biology 2012,
  16(5):284-287

To see these entries in BibTeX format, use 'print(&lt;citation&gt;,
bibtex=TRUE)', 'toBibtex(.)', or set
'options(citation.bibtex.max=999)'.</code></pre>
</div>
</div>


</section>

</main> <!-- /main -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const onCopySuccess = function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  }
  const getTextToCopy = function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button:not([data-in-quarto-modal])', {
    text: getTextToCopy
  });
  clipboard.on('success', onCopySuccess);
  if (window.document.getElementById('quarto-embedded-source-code-modal')) {
    // For code content inside modals, clipBoardJS needs to be initialized with a container option
    // TODO: Check when it could be a function (https://github.com/zenorocha/clipboard.js/issues/860)
    const clipboardModal = new window.ClipboardJS('.code-copy-button[data-in-quarto-modal]', {
      text: getTextToCopy,
      container: window.document.getElementById('quarto-embedded-source-code-modal')
    });
    clipboardModal.on('success', onCopySuccess);
  }
    var localhostRegex = new RegExp(/^(?:http|https):\/\/localhost\:?[0-9]*\//);
    var mailtoRegex = new RegExp(/^mailto:/);
      var filterRegex = new RegExp('/' + window.location.host + '/');
    var isInternal = (href) => {
        return filterRegex.test(href) || localhostRegex.test(href) || mailtoRegex.test(href);
    }
    // Inspect non-navigation links and adorn them if external
 	var links = window.document.querySelectorAll('a[href]:not(.nav-link):not(.navbar-brand):not(.toc-action):not(.sidebar-link):not(.sidebar-item-toggle):not(.pagination-link):not(.no-external):not([aria-hidden]):not(.dropdown-item):not(.quarto-navigation-tool):not(.about-link)');
    for (var i=0; i<links.length; i++) {
      const link = links[i];
      if (!isInternal(link.href)) {
        // undo the damage that might have been done by quarto-nav.js in the case of
        // links that we want to consider external
        if (link.dataset.originalHref !== undefined) {
          link.href = link.dataset.originalHref;
        }
      }
    }
  function tippyHover(el, contentFn, onTriggerFn, onUntriggerFn) {
    const config = {
      allowHTML: true,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start',
    };
    if (contentFn) {
      config.content = contentFn;
    }
    if (onTriggerFn) {
      config.onTrigger = onTriggerFn;
    }
    if (onUntriggerFn) {
      config.onUntrigger = onUntriggerFn;
    }
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      if (note) {
        return note.innerHTML;
      } else {
        return "";
      }
    });
  }
  const xrefs = window.document.querySelectorAll('a.quarto-xref');
  const processXRef = (id, note) => {
    // Strip column container classes
    const stripColumnClz = (el) => {
      el.classList.remove("page-full", "page-columns");
      if (el.children) {
        for (const child of el.children) {
          stripColumnClz(child);
        }
      }
    }
    stripColumnClz(note)
    if (id === null || id.startsWith('sec-')) {
      // Special case sections, only their first couple elements
      const container = document.createElement("div");
      if (note.children && note.children.length > 2) {
        container.appendChild(note.children[0].cloneNode(true));
        for (let i = 1; i < note.children.length; i++) {
          const child = note.children[i];
          if (child.tagName === "P" && child.innerText === "") {
            continue;
          } else {
            container.appendChild(child.cloneNode(true));
            break;
          }
        }
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(container);
        }
        return container.innerHTML
      } else {
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(note);
        }
        return note.innerHTML;
      }
    } else {
      // Remove any anchor links if they are present
      const anchorLink = note.querySelector('a.anchorjs-link');
      if (anchorLink) {
        anchorLink.remove();
      }
      if (window.Quarto?.typesetMath) {
        window.Quarto.typesetMath(note);
      }
      // TODO in 1.5, we should make sure this works without a callout special case
      if (note.classList.contains("callout")) {
        return note.outerHTML;
      } else {
        return note.innerHTML;
      }
    }
  }
  for (var i=0; i<xrefs.length; i++) {
    const xref = xrefs[i];
    tippyHover(xref, undefined, function(instance) {
      instance.disable();
      let url = xref.getAttribute('href');
      let hash = undefined; 
      if (url.startsWith('#')) {
        hash = url;
      } else {
        try { hash = new URL(url).hash; } catch {}
      }
      if (hash) {
        const id = hash.replace(/^#\/?/, "");
        const note = window.document.getElementById(id);
        if (note !== null) {
          try {
            const html = processXRef(id, note.cloneNode(true));
            instance.setContent(html);
          } finally {
            instance.enable();
            instance.show();
          }
        } else {
          // See if we can fetch this
          fetch(url.split('#')[0])
          .then(res => res.text())
          .then(html => {
            const parser = new DOMParser();
            const htmlDoc = parser.parseFromString(html, "text/html");
            const note = htmlDoc.getElementById(id);
            if (note !== null) {
              const html = processXRef(id, note);
              instance.setContent(html);
            } 
          }).finally(() => {
            instance.enable();
            instance.show();
          });
        }
      } else {
        // See if we can fetch a full url (with no hash to target)
        // This is a special case and we should probably do some content thinning / targeting
        fetch(url)
        .then(res => res.text())
        .then(html => {
          const parser = new DOMParser();
          const htmlDoc = parser.parseFromString(html, "text/html");
          const note = htmlDoc.querySelector('main.content');
          if (note !== null) {
            // This should only happen for chapter cross references
            // (since there is no id in the URL)
            // remove the first header
            if (note.children.length > 0 && note.children[0].tagName === "HEADER") {
              note.children[0].remove();
            }
            const html = processXRef(null, note);
            instance.setContent(html);
          } 
        }).finally(() => {
          instance.enable();
          instance.show();
        });
      }
    }, function(instance) {
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            div.style.left = 0;
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
        // Handle positioning of the toggle
    window.addEventListener(
      "resize",
      throttle(() => {
        elRect = undefined;
        if (selectedAnnoteEl) {
          selectCodeLines(selectedAnnoteEl);
        }
      }, 10)
    );
    function throttle(fn, ms) {
    let throttle = false;
    let timer;
      return (...args) => {
        if(!throttle) { // first call gets through
            fn.apply(this, args);
            throttle = true;
        } else { // all the others get throttled
            if(timer) clearTimeout(timer); // cancel #2
            timer = setTimeout(() => {
              fn.apply(this, args);
              timer = throttle = false;
            }, ms);
        }
      };
    }
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script>
</div> <!-- /content -->




</body></html>